# Analyse der Leistung von zwei Photovoltaikanlagen

## Projektziel

In diesem Projekt werden die Produktions- und Wetterdaten von zwei Photovoltaikanlagen untersucht.

Ziel ist es herauszufinden, wie Sonneneinstrahlung und Temperatur die Stromproduktion beeinflussen. Außerdem wird untersucht, ob einzelne Wechselrichter oder Zeiträume ungewöhnlich niedrige Leistungen aufweisen.

## Hauptfrage

Wie beeinflussen Sonneneinstrahlung und Temperatur die Stromproduktion der beiden PV-Anlagen, und lassen sich auffällige Leistungseinbußen erkennen?

## Hypothesen

1. Je höher die Sonneneinstrahlung ist, desto höher ist die erzeugte AC-Leistung.
2. Die Modultemperatur hängt stärker mit der Stromproduktion zusammen als die Außentemperatur.
3. Einzelne Wechselrichter erzeugen weniger Energie als die übrigen Wechselrichter derselben Anlage.
4. Anlage 2 weist häufiger auffällige Leistungseinbrüche auf als Anlage 1.
5. Die höchste Stromproduktion wird ungefähr zur Mittagszeit erreicht.

In [1]:
import pandas as pd
import plotly.express as px
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go



In [2]:
# Einheitliche Farben für alle Diagramme

FARBEN = {
    "Anlage 1": "#1F9E96",
    "Anlage 2": "#0B3C5D"
}

CSV-Dateien einlesen

In [3]:
# Produktionsdaten von Anlage 1 einlesen
gen1_df = pd.read_csv("Plant_1_Generation_Data.csv")

# Wetterdaten von Anlage 1 einlesen
wetter1_df = pd.read_csv("Plant_1_Weather_Sensor_Data.csv")

# Produktionsdaten von Anlage 2 einlesen
gen2_df = pd.read_csv("Plant_2_Generation_Data.csv")

# Wetterdaten von Anlage 2 einlesen
wetter2_df = pd.read_csv("Plant_2_Weather_Sensor_Data.csv")

Prüfen, ob alle Dateien geladen wurden

In [4]:
# Anzahl der Zeilen und Spalten anzeigen

print("Produktion Anlage 1:", gen1_df.shape)
print("Wetter Anlage 1:", wetter1_df.shape)
print("Produktion Anlage 2:", gen2_df.shape)
print("Wetter Anlage 2:", wetter2_df.shape)

Produktion Anlage 1: (68778, 7)
Wetter Anlage 1: (3182, 6)
Produktion Anlage 2: (67698, 7)
Wetter Anlage 2: (3259, 6)


In [5]:
# Erste fünf Zeilen der Produktionsdaten von Anlage 1 anzeigen

gen1_df.head()

,DATE_TIME,PLANT_ID,SOURCE_KEY,DC_POWER,AC_POWER,DAILY_YIELD,TOTAL_YIELD
0,15-05-2020 00:00,4135001,1BY6WEcLGh8j5v7,0.0,0.0,0.0,6259559.0
1,15-05-2020 00:00,4135001,1IF53ai7Xc0U56Y,0.0,0.0,0.0,6183645.0
2,15-05-2020 00:00,4135001,3PZuoBAID5Wc2HD,0.0,0.0,0.0,6987759.0
3,15-05-2020 00:00,4135001,7JYdWkrLSPkdwr4,0.0,0.0,0.0,7602960.0
4,15-05-2020 00:00,4135001,McdE0feGgRqW7Ca,0.0,0.0,0.0,7158964.0


In [6]:
# Erste fünf Zeilen der Wetterdaten von Anlage 1 anzeigen

wetter1_df.head()

,DATE_TIME,PLANT_ID,SOURCE_KEY,AMBIENT_TEMPERATURE,MODULE_TEMPERATURE,IRRADIATION
0,2020-05-15 00:00:00,4135001,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
1,2020-05-15 00:15:00,4135001,HmiyD2TTLFNqkNe,25.084589,22.761668,0.0
2,2020-05-15 00:30:00,4135001,HmiyD2TTLFNqkNe,24.935753,22.592306,0.0
3,2020-05-15 00:45:00,4135001,HmiyD2TTLFNqkNe,24.846130,22.360852,0.0
4,2020-05-15 01:00:00,4135001,HmiyD2TTLFNqkNe,24.621525,22.165423,0.0


In [7]:
# Erste fünf Zeilen der Produktionsdaten von Anlage 2 anzeigen

gen2_df.head()

,DATE_TIME,PLANT_ID,SOURCE_KEY,DC_POWER,AC_POWER,DAILY_YIELD,TOTAL_YIELD
0,2020-05-15 00:00:00,4136001,4UPUqMRk7TRMgml,0.0,0.0,9425.000000,2.429011e+06
1,2020-05-15 00:00:00,4136001,81aHJ1q11NBPMrL,0.0,0.0,0.000000,1.215279e+09
2,2020-05-15 00:00:00,4136001,9kRcWv60rDACzjR,0.0,0.0,3075.333333,2.247720e+09
3,2020-05-15 00:00:00,4136001,Et9kgGMDl729KT4,0.0,0.0,269.933333,1.704250e+06
4,2020-05-15 00:00:00,4136001,IQ2d7wF4YD8zU1Q,0.0,0.0,3177.000000,1.994153e+07


In [8]:
# Erste fünf Zeilen der Wetterdaten von Anlage 2 anzeigen

wetter2_df.head()

,DATE_TIME,PLANT_ID,SOURCE_KEY,AMBIENT_TEMPERATURE,MODULE_TEMPERATURE,IRRADIATION
0,2020-05-15 00:00:00,4136001,iq8k7ZNt4Mwm3w0,27.004764,25.060789,0.0
1,2020-05-15 00:15:00,4136001,iq8k7ZNt4Mwm3w0,26.880811,24.421869,0.0
2,2020-05-15 00:30:00,4136001,iq8k7ZNt4Mwm3w0,26.682055,24.427290,0.0
3,2020-05-15 00:45:00,4136001,iq8k7ZNt4Mwm3w0,26.500589,24.420678,0.0
4,2020-05-15 01:00:00,4136001,iq8k7ZNt4Mwm3w0,26.596148,25.088210,0.0


# Datenbereinigung

### Die Spalten in allen vier Tabellen einheitlich auf Deutsch umbenennen

In [9]:
# Spaltennamen für die Produktionsdaten festlegen

produktion_spalten = {
    "DATE_TIME": "datum",
    "PLANT_ID": "anlagen_id",
    "SOURCE_KEY": "wechselrichter_id",
    "DC_POWER": "dc_leistung",
    "AC_POWER": "ac_leistung",
    "DAILY_YIELD": "tagesertrag",
    "TOTAL_YIELD": "gesamtertrag"
}

In [10]:
# Spalten der Produktionsdaten umbenennen

gen1_df = gen1_df.rename(columns=produktion_spalten)
gen2_df = gen2_df.rename(columns=produktion_spalten)

In [11]:
# Neue Spaltennamen aller der Produktionstabellen anzeigen

print("Produktion Anlage 1:")
print(gen1_df.columns)

print("Produktion Anlage 2:")
print(gen2_df.columns)

Produktion Anlage 1:
Index(['datum', 'anlagen_id', 'wechselrichter_id', 'dc_leistung',
       'ac_leistung', 'tagesertrag', 'gesamtertrag'],
      dtype='str')
Produktion Anlage 2:
Index(['datum', 'anlagen_id', 'wechselrichter_id', 'dc_leistung',
       'ac_leistung', 'tagesertrag', 'gesamtertrag'],
      dtype='str')


In [12]:
gen1_df.head(3)

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag
0,15-05-2020 00:00,4135001,1BY6WEcLGh8j5v7,0.0,0.0,0.0,6259559.0
1,15-05-2020 00:00,4135001,1IF53ai7Xc0U56Y,0.0,0.0,0.0,6183645.0
2,15-05-2020 00:00,4135001,3PZuoBAID5Wc2HD,0.0,0.0,0.0,6987759.0


In [13]:
gen1_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 68778 entries, 0 to 68777
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   datum              68778 non-null  str    
 1   anlagen_id         68778 non-null  int64  
 2   wechselrichter_id  68778 non-null  str    
 3   dc_leistung        68778 non-null  float64
 4   ac_leistung        68778 non-null  float64
 5   tagesertrag        68778 non-null  float64
 6   gesamtertrag       68778 non-null  float64
dtypes: float64(4), int64(1), str(2)
memory usage: 3.7 MB


In [14]:
gen2_df.head(3)

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag
0,2020-05-15 00:00:00,4136001,4UPUqMRk7TRMgml,0.0,0.0,9425.000000,2.429011e+06
1,2020-05-15 00:00:00,4136001,81aHJ1q11NBPMrL,0.0,0.0,0.000000,1.215279e+09
2,2020-05-15 00:00:00,4136001,9kRcWv60rDACzjR,0.0,0.0,3075.333333,2.247720e+09


In [15]:
gen2_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 67698 entries, 0 to 67697
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   datum              67698 non-null  str    
 1   anlagen_id         67698 non-null  int64  
 2   wechselrichter_id  67698 non-null  str    
 3   dc_leistung        67698 non-null  float64
 4   ac_leistung        67698 non-null  float64
 5   tagesertrag        67698 non-null  float64
 6   gesamtertrag       67698 non-null  float64
dtypes: float64(4), int64(1), str(2)
memory usage: 3.6 MB


In [16]:
# Spaltennamen für die Wetterdaten festlegen

wetter_spalten = {
    "DATE_TIME": "datum",
    "PLANT_ID": "anlagen_id",
    "SOURCE_KEY": "sensor_id",
    "AMBIENT_TEMPERATURE": "aussentemperatur",
    "MODULE_TEMPERATURE": "modultemperatur",
    "IRRADIATION": "sonneneinstrahlung"
}

In [17]:
# Spalten der Wetterdaten umbenennen

wetter1_df = wetter1_df.rename(columns=wetter_spalten)
wetter2_df = wetter2_df.rename(columns=wetter_spalten)

In [18]:
# Neue Spaltennamen aller der Wettertabellen anzeigen#

print("Wetter Anlage 1:")
print(wetter1_df.columns)


print("Wetter Anlage 2:")
print(wetter2_df.columns)

Wetter Anlage 1:
Index(['datum', 'anlagen_id', 'sensor_id', 'aussentemperatur',
       'modultemperatur', 'sonneneinstrahlung'],
      dtype='str')
Wetter Anlage 2:
Index(['datum', 'anlagen_id', 'sensor_id', 'aussentemperatur',
       'modultemperatur', 'sonneneinstrahlung'],
      dtype='str')


In [19]:
wetter1_df.head(3)

,datum,anlagen_id,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
0,2020-05-15 00:00:00,4135001,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
1,2020-05-15 00:15:00,4135001,HmiyD2TTLFNqkNe,25.084589,22.761668,0.0
2,2020-05-15 00:30:00,4135001,HmiyD2TTLFNqkNe,24.935753,22.592306,0.0


In [20]:
wetter1_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3182 entries, 0 to 3181
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   datum               3182 non-null   str    
 1   anlagen_id          3182 non-null   int64  
 2   sensor_id           3182 non-null   str    
 3   aussentemperatur    3182 non-null   float64
 4   modultemperatur     3182 non-null   float64
 5   sonneneinstrahlung  3182 non-null   float64
dtypes: float64(3), int64(1), str(2)
memory usage: 149.3 KB


In [21]:
wetter2_df.head(3)

,datum,anlagen_id,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
0,2020-05-15 00:00:00,4136001,iq8k7ZNt4Mwm3w0,27.004764,25.060789,0.0
1,2020-05-15 00:15:00,4136001,iq8k7ZNt4Mwm3w0,26.880811,24.421869,0.0
2,2020-05-15 00:30:00,4136001,iq8k7ZNt4Mwm3w0,26.682055,24.427290,0.0


In [22]:
wetter2_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3259 entries, 0 to 3258
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   datum               3259 non-null   str    
 1   anlagen_id          3259 non-null   int64  
 2   sensor_id           3259 non-null   str    
 3   aussentemperatur    3259 non-null   float64
 4   modultemperatur     3259 non-null   float64
 5   sonneneinstrahlung  3259 non-null   float64
dtypes: float64(3), int64(1), str(2)
memory usage: 152.9 KB


In [23]:
# Datum in den Produktionsdaten von Anlage 1 umwandeln

gen1_df["datum"] = pd.to_datetime(gen1_df["datum"],format="%d-%m-%Y %H:%M")
gen1_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 68778 entries, 0 to 68777
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   datum              68778 non-null  datetime64[us]
 1   anlagen_id         68778 non-null  int64         
 2   wechselrichter_id  68778 non-null  str           
 3   dc_leistung        68778 non-null  float64       
 4   ac_leistung        68778 non-null  float64       
 5   tagesertrag        68778 non-null  float64       
 6   gesamtertrag       68778 non-null  float64       
dtypes: datetime64[us](1), float64(4), int64(1), str(1)
memory usage: 3.7 MB


In [24]:
gen1_df.head(20)

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag
0,2020-05-15,4135001,1BY6WEcLGh8j5v7,0.0,0.0,0.0,6259559.0
1,2020-05-15,4135001,1IF53ai7Xc0U56Y,0.0,0.0,0.0,6183645.0
2,2020-05-15,4135001,3PZuoBAID5Wc2HD,0.0,0.0,0.0,6987759.0
3,2020-05-15,4135001,7JYdWkrLSPkdwr4,0.0,0.0,0.0,7602960.0
4,2020-05-15,4135001,McdE0feGgRqW7Ca,0.0,0.0,0.0,7158964.0
5,2020-05-15,4135001,VHMLBKoKgIrUVDU,0.0,0.0,0.0,7206408.0
6,2020-05-15,4135001,WRmjgnKYAwPKWDb,0.0,0.0,0.0,7028673.0
7,2020-05-15,4135001,ZnxXDlPa8U1GXgE,0.0,0.0,0.0,6522172.0
8,2020-05-15,4135001,ZoEaEvLYb1n2sOq,0.0,0.0,0.0,7098099.0
9,2020-05-15,4135001,adLQvlD726eNBSB,0.0,0.0,0.0,6271355.0


In [25]:
# Datum in den Produktionsdaten von Anlage 2 umwandeln

gen2_df["datum"] = pd.to_datetime(gen2_df["datum"],format="%Y-%m-%d %H:%M:%S")
gen2_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 67698 entries, 0 to 67697
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   datum              67698 non-null  datetime64[us]
 1   anlagen_id         67698 non-null  int64         
 2   wechselrichter_id  67698 non-null  str           
 3   dc_leistung        67698 non-null  float64       
 4   ac_leistung        67698 non-null  float64       
 5   tagesertrag        67698 non-null  float64       
 6   gesamtertrag       67698 non-null  float64       
dtypes: datetime64[us](1), float64(4), int64(1), str(1)
memory usage: 3.6 MB


In [26]:
gen2_df.head()

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag
0,2020-05-15,4136001,4UPUqMRk7TRMgml,0.0,0.0,9425.000000,2.429011e+06
1,2020-05-15,4136001,81aHJ1q11NBPMrL,0.0,0.0,0.000000,1.215279e+09
2,2020-05-15,4136001,9kRcWv60rDACzjR,0.0,0.0,3075.333333,2.247720e+09
3,2020-05-15,4136001,Et9kgGMDl729KT4,0.0,0.0,269.933333,1.704250e+06
4,2020-05-15,4136001,IQ2d7wF4YD8zU1Q,0.0,0.0,3177.000000,1.994153e+07


In [27]:
# Datum Wettertabelle 1 umwandeln

wetter1_df["datum"] = pd.to_datetime(wetter1_df["datum"],format="%Y-%m-%d %H:%M:%S")
wetter1_df.info()


<class 'pandas.DataFrame'>
RangeIndex: 3182 entries, 0 to 3181
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   datum               3182 non-null   datetime64[us]
 1   anlagen_id          3182 non-null   int64         
 2   sensor_id           3182 non-null   str           
 3   aussentemperatur    3182 non-null   float64       
 4   modultemperatur     3182 non-null   float64       
 5   sonneneinstrahlung  3182 non-null   float64       
dtypes: datetime64[us](1), float64(3), int64(1), str(1)
memory usage: 149.3 KB


In [28]:
wetter1_df.head()

,datum,anlagen_id,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
0,2020-05-15 00:00:00,4135001,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
1,2020-05-15 00:15:00,4135001,HmiyD2TTLFNqkNe,25.084589,22.761668,0.0
2,2020-05-15 00:30:00,4135001,HmiyD2TTLFNqkNe,24.935753,22.592306,0.0
3,2020-05-15 00:45:00,4135001,HmiyD2TTLFNqkNe,24.846130,22.360852,0.0
4,2020-05-15 01:00:00,4135001,HmiyD2TTLFNqkNe,24.621525,22.165423,0.0


In [29]:
wetter2_df["datum"] = pd.to_datetime(wetter2_df["datum"],format="%Y-%m-%d %H:%M:%S")

wetter2_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3259 entries, 0 to 3258
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   datum               3259 non-null   datetime64[us]
 1   anlagen_id          3259 non-null   int64         
 2   sensor_id           3259 non-null   str           
 3   aussentemperatur    3259 non-null   float64       
 4   modultemperatur     3259 non-null   float64       
 5   sonneneinstrahlung  3259 non-null   float64       
dtypes: datetime64[us](1), float64(3), int64(1), str(1)
memory usage: 152.9 KB


In [30]:
wetter2_df.head()

,datum,anlagen_id,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
0,2020-05-15 00:00:00,4136001,iq8k7ZNt4Mwm3w0,27.004764,25.060789,0.0
1,2020-05-15 00:15:00,4136001,iq8k7ZNt4Mwm3w0,26.880811,24.421869,0.0
2,2020-05-15 00:30:00,4136001,iq8k7ZNt4Mwm3w0,26.682055,24.427290,0.0
3,2020-05-15 00:45:00,4136001,iq8k7ZNt4Mwm3w0,26.500589,24.420678,0.0
4,2020-05-15 01:00:00,4136001,iq8k7ZNt4Mwm3w0,26.596148,25.088210,0.0


In [31]:
# Datentyp der Datumsspalte kontrollieren

print("Produktion Anlage 1:", gen1_df["datum"].dtype)
print("Produktion Anlage 2:", gen2_df["datum"].dtype)
print("Wetter Anlage 1:", wetter1_df["datum"].dtype)
print("Wetter Anlage 2:", wetter2_df["datum"].dtype)

Produktion Anlage 1: datetime64[us]
Produktion Anlage 2: datetime64[us]
Wetter Anlage 1: datetime64[us]
Wetter Anlage 2: datetime64[us]


In [32]:
# Frühestes und spätestes Datum anzeigen

print("Beginn:", gen1_df["datum"].min())
print("Ende:", gen1_df["datum"].max())

Beginn: 2020-05-15 00:00:00
Ende: 2020-06-17 23:45:00


## Fehlende Werte kontrollieren

In [33]:
# Fehlende Werte in allen vier Tabellen zählen

print("Fehlende Werte - Produktion Anlage 1")
print(gen1_df.isna().sum())

print("Fehlende Werte - Wetter Anlage 1")
print(wetter1_df.isna().sum())

print("Fehlende Werte - Produktion Anlage 2")
print(gen2_df.isna().sum())

print("Fehlende Werte - Wetter Anlage 2")
print(wetter2_df.isna().sum())

Fehlende Werte - Produktion Anlage 1
datum                0
anlagen_id           0
wechselrichter_id    0
dc_leistung          0
ac_leistung          0
tagesertrag          0
gesamtertrag         0
dtype: int64
Fehlende Werte - Wetter Anlage 1
datum                 0
anlagen_id            0
sensor_id             0
aussentemperatur      0
modultemperatur       0
sonneneinstrahlung    0
dtype: int64
Fehlende Werte - Produktion Anlage 2
datum                0
anlagen_id           0
wechselrichter_id    0
dc_leistung          0
ac_leistung          0
tagesertrag          0
gesamtertrag         0
dtype: int64
Fehlende Werte - Wetter Anlage 2
datum                 0
anlagen_id            0
sensor_id             0
aussentemperatur      0
modultemperatur       0
sonneneinstrahlung    0
dtype: int64


## Duplikate kontrollieren

In [34]:
# Vollständig doppelte Zeilen zählen

print("Duplikate Produktion Anlage 1:", gen1_df.duplicated().sum())
print("Duplikate Wetter Anlage 1:", wetter1_df.duplicated().sum())
print("Duplikate Produktion Anlage 2:", gen2_df.duplicated().sum())
print("Duplikate Wetter Anlage 2:", wetter2_df.duplicated().sum())

Duplikate Produktion Anlage 1: 0
Duplikate Wetter Anlage 1: 0
Duplikate Produktion Anlage 2: 0
Duplikate Wetter Anlage 2: 0


## Anlagenbezeichnung ergänzen

In [35]:
# Verständliche Bezeichnung der Anlagen ergänzen

gen1_df["anlage"] = "Anlage 1"
gen2_df["anlage"] = "Anlage 2"

wetter1_df["anlage"] = "Anlage 1"
wetter2_df["anlage"] = "Anlage 2"

In [36]:
# Kontrolle
gen1_df.head(3)

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag,anlage
0,2020-05-15,4135001,1BY6WEcLGh8j5v7,0.0,0.0,0.0,6259559.0,Anlage 1
1,2020-05-15,4135001,1IF53ai7Xc0U56Y,0.0,0.0,0.0,6183645.0,Anlage 1
2,2020-05-15,4135001,3PZuoBAID5Wc2HD,0.0,0.0,0.0,6987759.0,Anlage 1


## Produktionsdaten zusammenfügen

In [37]:
# Produktionsdaten beider Anlagen untereinander verbinden

produktion_df = pd.concat([gen1_df, gen2_df], ignore_index= True)

In [38]:
# Kontrolle

print(produktion_df.shape)
produktion_df.head()

(136476, 8)


,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag,anlage
0,2020-05-15,4135001,1BY6WEcLGh8j5v7,0.0,0.0,0.0,6259559.0,Anlage 1
1,2020-05-15,4135001,1IF53ai7Xc0U56Y,0.0,0.0,0.0,6183645.0,Anlage 1
2,2020-05-15,4135001,3PZuoBAID5Wc2HD,0.0,0.0,0.0,6987759.0,Anlage 1
3,2020-05-15,4135001,7JYdWkrLSPkdwr4,0.0,0.0,0.0,7602960.0,Anlage 1
4,2020-05-15,4135001,McdE0feGgRqW7Ca,0.0,0.0,0.0,7158964.0,Anlage 1


## Wetterdaten zusammenfügen

In [39]:
# Wetterdaten beider Anlagen untereinander verbinden

wetter_df = pd.concat([wetter1_df, wetter2_df], ignore_index= True)

In [40]:
# Kontrolle

print(wetter_df.shape)
wetter_df.head()

(6441, 7)


,datum,anlagen_id,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung,anlage
0,2020-05-15 00:00:00,4135001,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0,Anlage 1
1,2020-05-15 00:15:00,4135001,HmiyD2TTLFNqkNe,25.084589,22.761668,0.0,Anlage 1
2,2020-05-15 00:30:00,4135001,HmiyD2TTLFNqkNe,24.935753,22.592306,0.0,Anlage 1
3,2020-05-15 00:45:00,4135001,HmiyD2TTLFNqkNe,24.846130,22.360852,0.0,Anlage 1
4,2020-05-15 01:00:00,4135001,HmiyD2TTLFNqkNe,24.621525,22.165423,0.0,Anlage 1


## Prüfen, ob beide Anlagen enthalten sind

In [41]:
# Anzahl der Datensätze je Anlage anzeigen

produktion_df["anlage"].value_counts()

anlage
Anlage 1    68778
Anlage 2    67698
Name: count, dtype: int64

In [42]:
# Anzahl der Datensätze je Wettertabelle 
wetter_df["anlage"].value_counts()

anlage
Anlage 2    3259
Anlage 1    3182
Name: count, dtype: int64

## Anzahl der Wechselrichter prüfen

In [43]:
# Anzahl unterschiedlicher Wechselrichter je Anlage

produktion_df.groupby("anlage")["wechselrichter_id"].nunique()

anlage
Anlage 1    22
Anlage 2    22
Name: wechselrichter_id, dtype: int64

## Schlüssel vor dem Merge prüfen

In [44]:
# Doppelte Wetterdaten pro Anlage und Zeitpunkt prüfen

wetter_df.duplicated(subset=["datum", "anlagen_id"]).sum()

np.int64(0)

## Produktion und Wetter verbinden

In [45]:
# Produktions- und Wetterdaten über Datum und Anlagen-ID verbinden

analyse_df = produktion_df.merge(wetter_df,on=["datum", "anlagen_id", "anlage"],how="left")

In [46]:
# Größe der verbundenen Tabelle prüfen

print(analyse_df.shape)
analyse_df.head()

(136476, 12)


,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag,anlage,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
0,2020-05-15,4135001,1BY6WEcLGh8j5v7,0.0,0.0,0.0,6259559.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
1,2020-05-15,4135001,1IF53ai7Xc0U56Y,0.0,0.0,0.0,6183645.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
2,2020-05-15,4135001,3PZuoBAID5Wc2HD,0.0,0.0,0.0,6987759.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
3,2020-05-15,4135001,7JYdWkrLSPkdwr4,0.0,0.0,0.0,7602960.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
4,2020-05-15,4135001,McdE0feGgRqW7Ca,0.0,0.0,0.0,7158964.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0


In [47]:
# Fehlende Werte in der verbundenen Tabelle zählen

analyse_df.isna().sum()

datum                 0
anlagen_id            0
wechselrichter_id     0
dc_leistung           0
ac_leistung           0
tagesertrag           0
gesamtertrag          0
anlage                0
sensor_id             4
aussentemperatur      4
modultemperatur       4
sonneneinstrahlung    4
dtype: int64

In [48]:
analyse_df.shape

(136476, 12)

In [49]:
# 1. Zeilen rund um die Stelle VOR dem Löschen ansehen
display(analyse_df.loc[38542:38549])

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag,anlage,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
38542,2020-06-03 13:45:00,4135001,zBIq5rxdHJRwDNY,8344.50000,816.650000,5016.000000,6481470.000,Anlage 1,HmiyD2TTLFNqkNe,27.862188,46.631439,0.615273
38543,2020-06-03 13:45:00,4135001,zVJPv84UY57bAof,9345.50000,914.100000,4998.000000,7260215.000,Anlage 1,HmiyD2TTLFNqkNe,27.862188,46.631439,0.615273
38544,2020-06-03 14:00:00,4135001,1IF53ai7Xc0U56Y,7003.00000,685.800000,5601.000000,6330385.000,Anlage 1,NaN,NaN,NaN,NaN
38545,2020-06-03 14:00:00,4135001,adLQvlD726eNBSB,7204.00000,705.400000,5685.000000,6419961.000,Anlage 1,NaN,NaN,NaN,NaN
38546,2020-06-03 14:00:00,4135001,wCURE6d3bPkepu2,7545.00000,738.700000,5579.000000,6928448.000,Anlage 1,NaN,NaN,NaN,NaN
38547,2020-06-03 14:00:00,4135001,z9Y9gH1T5YWrNuG,7946.00000,777.800000,5541.000000,7152815.000,Anlage 1,NaN,NaN,NaN,NaN
38548,2020-06-03 14:15:00,4135001,1BY6WEcLGh8j5v7,8785.87500,859.550000,5306.250000,6395774.250,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680
38549,2020-06-03 14:15:00,4135001,1IF53ai7Xc0U56Y,10615.71429,1037.357143,5726.714286,6330510.714,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680


In [50]:
# Zeilen mit fehlenden Wetterdaten anzeigen

analyse_df[analyse_df["sonneneinstrahlung"].isna()]

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag,anlage,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
38544,2020-06-03 14:00:00,4135001,1IF53ai7Xc0U56Y,7003.0,685.8,5601.0,6330385.0,Anlage 1,NaN,NaN,NaN,NaN
38545,2020-06-03 14:00:00,4135001,adLQvlD726eNBSB,7204.0,705.4,5685.0,6419961.0,Anlage 1,NaN,NaN,NaN,NaN
38546,2020-06-03 14:00:00,4135001,wCURE6d3bPkepu2,7545.0,738.7,5579.0,6928448.0,Anlage 1,NaN,NaN,NaN,NaN
38547,2020-06-03 14:00:00,4135001,z9Y9gH1T5YWrNuG,7946.0,777.8,5541.0,7152815.0,Anlage 1,NaN,NaN,NaN,NaN


In [51]:
# Zeilen mit fehlenden Wetterdaten entfernen, da es nur 4 von 136.476 Zeilen sind

analyse_df = analyse_df.dropna(subset=["sensor_id","aussentemperatur","modultemperatur","sonneneinstrahlung"])
analyse_df.shape

(136472, 12)

In [52]:
analyse_df.head(80)

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag,anlage,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
0,2020-05-15 00:00:00,4135001,1BY6WEcLGh8j5v7,0.0,0.0,0.0,6259559.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
1,2020-05-15 00:00:00,4135001,1IF53ai7Xc0U56Y,0.0,0.0,0.0,6183645.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
2,2020-05-15 00:00:00,4135001,3PZuoBAID5Wc2HD,0.0,0.0,0.0,6987759.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
3,2020-05-15 00:00:00,4135001,7JYdWkrLSPkdwr4,0.0,0.0,0.0,7602960.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
4,2020-05-15 00:00:00,4135001,McdE0feGgRqW7Ca,0.0,0.0,0.0,7158964.0,Anlage 1,HmiyD2TTLFNqkNe,25.184316,22.857507,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...
75,2020-05-15 00:45:00,4135001,ih0vzX44oOqAx2f,0.0,0.0,0.0,6185184.0,Anlage 1,HmiyD2TTLFNqkNe,24.846130,22.360852,0.0
76,2020-05-15 00:45:00,4135001,pkci93gMrogZuBj,0.0,0.0,0.0,7169102.0,Anlage 1,HmiyD2TTLFNqkNe,24.846130,22.360852,0.0
77,2020-05-15 00:45:00,4135001,rGa61gmuvPhdLxV,0.0,0.0,0.0,7111493.0,Anlage 1,HmiyD2TTLFNqkNe,24.846130,22.360852,0.0
78,2020-05-15 00:45:00,4135001,sjndEbLyjtCKgGv,0.0,0.0,0.0,7016832.0,Anlage 1,HmiyD2TTLFNqkNe,24.846130,22.360852,0.0


In [53]:
# 1. Zeilen rund um die Stelle VOR dem Löschen ansehen
display(analyse_df.loc[38542:38549])

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag,anlage,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
38542,2020-06-03 13:45:00,4135001,zBIq5rxdHJRwDNY,8344.50000,816.650000,5016.000000,6481470.000,Anlage 1,HmiyD2TTLFNqkNe,27.862188,46.631439,0.615273
38543,2020-06-03 13:45:00,4135001,zVJPv84UY57bAof,9345.50000,914.100000,4998.000000,7260215.000,Anlage 1,HmiyD2TTLFNqkNe,27.862188,46.631439,0.615273
38548,2020-06-03 14:15:00,4135001,1BY6WEcLGh8j5v7,8785.87500,859.550000,5306.250000,6395774.250,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680
38549,2020-06-03 14:15:00,4135001,1IF53ai7Xc0U56Y,10615.71429,1037.357143,5726.714286,6330510.714,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680


In [54]:
# Index nach dem Löschen neu erstellen

analyse_df = analyse_df.reset_index(drop=True)


display(analyse_df.loc[38542:38549])

,datum,anlagen_id,wechselrichter_id,dc_leistung,ac_leistung,tagesertrag,gesamtertrag,anlage,sensor_id,aussentemperatur,modultemperatur,sonneneinstrahlung
38542,2020-06-03 13:45:00,4135001,zBIq5rxdHJRwDNY,8344.50000,816.650000,5016.000000,6481470.000,Anlage 1,HmiyD2TTLFNqkNe,27.862188,46.631439,0.615273
38543,2020-06-03 13:45:00,4135001,zVJPv84UY57bAof,9345.50000,914.100000,4998.000000,7260215.000,Anlage 1,HmiyD2TTLFNqkNe,27.862188,46.631439,0.615273
38544,2020-06-03 14:15:00,4135001,1BY6WEcLGh8j5v7,8785.87500,859.550000,5306.250000,6395774.250,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680
38545,2020-06-03 14:15:00,4135001,1IF53ai7Xc0U56Y,10615.71429,1037.357143,5726.714286,6330510.714,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680
38546,2020-06-03 14:15:00,4135001,3PZuoBAID5Wc2HD,9945.25000,972.087500,5726.125000,7134340.125,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680
38547,2020-06-03 14:15:00,4135001,7JYdWkrLSPkdwr4,9946.50000,972.137500,5612.750000,7746485.750,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680
38548,2020-06-03 14:15:00,4135001,McdE0feGgRqW7Ca,10188.57143,995.614286,5769.428571,7306128.429,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680
38549,2020-06-03 14:15:00,4135001,VHMLBKoKgIrUVDU,9781.12500,956.137500,5744.625000,7353488.625,Anlage 1,HmiyD2TTLFNqkNe,28.361993,50.627043,0.716680


In [55]:
# Größe und fehlende Werte erneut prüfen

print(analyse_df.shape)
print(analyse_df.isna().sum())

(136472, 12)
datum                 0
anlagen_id            0
wechselrichter_id     0
dc_leistung           0
ac_leistung           0
tagesertrag           0
gesamtertrag          0
anlage                0
sensor_id             0
aussentemperatur      0
modultemperatur       0
sonneneinstrahlung    0
dtype: int64


Nach dem Zusammenführen fehlten bei vier Produktionszeilen die zugehörigen Wetterdaten. Da nur 4 von insgesamt 136.476 Zeilen betroffen waren, wurden diese Zeilen entfernt. Dadurch entsteht kein bedeutender Datenverlust.

## Zahlenwerte beschreiben

In [56]:
# Statistische Übersicht der wichtigsten Messwerte

analyse_df[[
    "dc_leistung",
    "ac_leistung",
    "tagesertrag",
    "gesamtertrag",
    "aussentemperatur",
    "modultemperatur",
    "sonneneinstrahlung"
]].describe().round(2)

,dc_leistung,ac_leistung,tagesertrag,gesamtertrag,aussentemperatur,modultemperatur,sonneneinstrahlung
count,136472.00,136472.00,136472.00,1.364720e+05,136472.00,136472.00,136472.00
mean,1708.37,274.79,3295.37,3.303916e+08,26.76,31.92,0.23
std,3222.08,380.18,3035.31,6.085769e+08,3.90,11.80,0.31
min,0.00,0.00,0.00,0.000000e+00,20.40,18.14,0.00
25%,0.00,0.00,28.29,6.520020e+06,23.64,22.41,0.00
50%,5.99,3.49,2834.64,7.269333e+06,25.91,26.41,0.03
75%,1155.59,532.57,5992.00,2.826096e+08,29.27,40.78,0.44
max,14471.12,1410.95,9873.00,2.247916e+09,39.18,66.64,1.22


## Datenqualität der DC-Werte prüfen

Vor der eigentlichen Analyse wird kontrolliert, wie sich die DC-Leistung und die AC-Leistung der beiden Anlagen zueinander verhalten.

Ein ungewöhnlich großes Verhältnis kann auf unterschiedliche Skalierungen oder Einheiten in den Ausgangsdaten hinweisen.

In [57]:
# Gesamte AC- und DC-Leistung je Anlage berechnen

ac_dc_df = (
    analyse_df
    .groupby("anlage", as_index=False)
    .agg(
        ac_leistung_gesamt=("ac_leistung", "sum"),
        dc_leistung_gesamt=("dc_leistung", "sum")
    )
)

# Verhältnis von AC-Leistung zu DC-Leistung berechnen

ac_dc_df["ac_dc_verhaeltnis"] = (
    ac_dc_df["ac_leistung_gesamt"]
    / ac_dc_df["dc_leistung_gesamt"]
)

# Verhältnis als Prozentwert anzeigen

ac_dc_df["ac_dc_prozent"] = (
    ac_dc_df["ac_dc_verhaeltnis"] * 100
).round(1)

ac_dc_df

,anlage,ac_leistung_gesamt,dc_leistung_gesamt,ac_dc_verhaeltnis,ac_dc_prozent
0,Anlage 1,2.116715e+07,2.164440e+08,0.097795,9.8
1,Anlage 2,1.633403e+07,1.670123e+07,0.978013,97.8


### Ergebnis der Datenqualitätsprüfung

Bei Anlage 1 beträgt das Verhältnis der gesamten AC-Leistung zur DC-Leistung nur ungefähr 9,8 %. Bei Anlage 2 liegt das Verhältnis bei ungefähr 97,8 %.

Die DC-Werte von Anlage 1 scheinen daher anders skaliert oder in einer anderen Einheit gespeichert zu sein. Die genaue Ursache kann mit den vorhandenen Informationen nicht bestimmt werden.

Für den Vergleich der beiden Anlagen werden deshalb hauptsächlich die AC-Leistung und relative Werte verwendet. Die absoluten DC-Werte der beiden Anlagen werden nicht direkt miteinander verglichen.

## Negative Werte kontrollieren

In [58]:
# Negative Werte in den Messspalten zählen

messspalten = ["dc_leistung","ac_leistung","tagesertrag","gesamtertrag","sonneneinstrahlung"]

(analyse_df[messspalten] < 0).sum()

dc_leistung           0
ac_leistung           0
tagesertrag           0
gesamtertrag          0
sonneneinstrahlung    0
dtype: int64

## Neue Zeitspalten erstellen

In [59]:
# Datum ohne Uhrzeit erstellen

analyse_df["tag"] = analyse_df["datum"].dt.date

In [60]:
# Stunde aus der Datumsspalte übernehmen

analyse_df["stunde"] = analyse_df["datum"].dt.hour

In [61]:
# Wochentag erstellen

analyse_df["wochentag"] = analyse_df["datum"].dt.day_name()

In [62]:
# Ergebnis kontrollieren

analyse_df[["datum","tag","stunde","wochentag"]].head()

,datum,tag,stunde,wochentag
0,2020-05-15,2020-05-15,0,Friday
1,2020-05-15,2020-05-15,0,Friday
2,2020-05-15,2020-05-15,0,Friday
3,2020-05-15,2020-05-15,0,Friday
4,2020-05-15,2020-05-15,0,Friday


## Produktionswerte je Zeitpunkt zusammenfassen

In [63]:
# Messwerte aller Wechselrichter je Anlage und Zeitpunkt zusammenfassen

zeitpunkt_df = analyse_df.groupby(["anlage", "datum"],as_index=False).agg({"dc_leistung": "sum","ac_leistung": "sum","aussentemperatur": "first","modultemperatur": "first","sonneneinstrahlung": "first"})

zeitpunkt_df.head() # um die Leistungen zu Checken besser head.(60) anzuwenden

,anlage,datum,dc_leistung,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung
0,Anlage 1,2020-05-15 00:00:00,0.0,0.0,25.184316,22.857507,0.0
1,Anlage 1,2020-05-15 00:15:00,0.0,0.0,25.084589,22.761668,0.0
2,Anlage 1,2020-05-15 00:30:00,0.0,0.0,24.935753,22.592306,0.0
3,Anlage 1,2020-05-15 00:45:00,0.0,0.0,24.846130,22.360852,0.0
4,Anlage 1,2020-05-15 01:00:00,0.0,0.0,24.621525,22.165423,0.0


In [64]:
# Wie viele Wechselrichter haben pro Zeitpunkt Daten geliefert?
melder = (analyse_df.groupby(["anlage", "datum"])["wechselrichter_id"]
          .nunique().rename("anzahl_wr").reset_index())
zeitpunkt_df = zeitpunkt_df.merge(melder, on=["anlage", "datum"])

# Mittlere AC-Leistung je meldendem Wechselrichter
zeitpunkt_df["ac_je_wr"] = zeitpunkt_df["ac_leistung"] / zeitpunkt_df["anzahl_wr"]
zeitpunkt_df["ac_je_wr_relativ"] = (zeitpunkt_df["ac_je_wr"]
    / zeitpunkt_df.groupby("anlage")["ac_je_wr"].transform("max"))

# Anteil vollständiger Zeitpunkte am Tag (Einstrahlung > 0,1)
tag = zeitpunkt_df[zeitpunkt_df["sonneneinstrahlung"] > 0.1]
print((tag["anzahl_wr"] == 22).groupby(tag["anlage"]).mean().round(3))

anlage
Anlage 1    0.997
Anlage 2    0.707
Name: anzahl_wr, dtype: float64


### Datenqualität: Anzahl der meldenden Wechselrichter

Die bisherige Berechnung verwendet die Summe der AC-Leistung aller Wechselrichter je Anlage und Zeitpunkt. Wenn einzelne Wechselrichter zu einem Zeitpunkt keine Messwerte liefern, fällt diese Summe niedriger aus. Dadurch kann ein Leistungseinbruch entstehen, obwohl lediglich Messwerte fehlen.

Bei Anlage 1 melden bei 99,7 % der untersuchten Tageszeitpunkte alle 22 Wechselrichter. Bei Anlage 2 trifft dies nur auf 70,7 % der Tageszeitpunkte zu.

Deshalb wurde zusätzlich die durchschnittliche AC-Leistung je meldendem Wechselrichter berechnet. Nach dieser Anpassung sinkt die Zahl der auffälligen Messzeitpunkte bei Anlage 2 von 30 auf 19. Bei Anlage 1 werden weiterhin keine Auffälligkeiten erkannt.

Die Korrelation zwischen Sonneneinstrahlung und AC-Leistung steigt bei Anlage 2 von 0,91 auf 0,93.

Die Kernaussage der Analyse bleibt damit bestehen: Anlage 2 weist auch nach Berücksichtigung der fehlenden Meldungen häufiger auffällige Leistungseinbrüche auf als Anlage 1. Die ursprüngliche Anzahl von 30 Messzeitpunkten wurde jedoch teilweise durch Datenlücken beeinflusst.

In [65]:
# Größe und erste Zeilen der neuen Tabelle anzeigen

print(zeitpunkt_df.shape)
zeitpunkt_df.head()

(6416, 10)


,anlage,datum,dc_leistung,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung,anzahl_wr,ac_je_wr,ac_je_wr_relativ
0,Anlage 1,2020-05-15 00:00:00,0.0,0.0,25.184316,22.857507,0.0,21,0.0,0.0
1,Anlage 1,2020-05-15 00:15:00,0.0,0.0,25.084589,22.761668,0.0,21,0.0,0.0
2,Anlage 1,2020-05-15 00:30:00,0.0,0.0,24.935753,22.592306,0.0,21,0.0,0.0
3,Anlage 1,2020-05-15 00:45:00,0.0,0.0,24.846130,22.360852,0.0,21,0.0,0.0
4,Anlage 1,2020-05-15 01:00:00,0.0,0.0,24.621525,22.165423,0.0,22,0.0,0.0


## Neue Zeitspalten ergänzen


In [66]:
# Zeitspalten für spätere Analysen erstellen

zeitpunkt_df["tag"] = zeitpunkt_df["datum"].dt.date
zeitpunkt_df["stunde"] = zeitpunkt_df["datum"].dt.hour

zeitpunkt_df.head(20)

,anlage,datum,dc_leistung,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung,anzahl_wr,ac_je_wr,ac_je_wr_relativ,tag,stunde
0,Anlage 1,2020-05-15 00:00:00,0.0,0.0,25.184316,22.857507,0.0,21,0.0,0.0,2020-05-15,0
1,Anlage 1,2020-05-15 00:15:00,0.0,0.0,25.084589,22.761668,0.0,21,0.0,0.0,2020-05-15,0
2,Anlage 1,2020-05-15 00:30:00,0.0,0.0,24.935753,22.592306,0.0,21,0.0,0.0,2020-05-15,0
3,Anlage 1,2020-05-15 00:45:00,0.0,0.0,24.846130,22.360852,0.0,21,0.0,0.0,2020-05-15,0
4,Anlage 1,2020-05-15 01:00:00,0.0,0.0,24.621525,22.165423,0.0,22,0.0,0.0,2020-05-15,1
5,Anlage 1,2020-05-15 01:15:00,0.0,0.0,24.536092,21.968571,0.0,21,0.0,0.0,2020-05-15,1
6,Anlage 1,2020-05-15 01:30:00,0.0,0.0,24.638674,22.352926,0.0,21,0.0,0.0,2020-05-15,1
7,Anlage 1,2020-05-15 01:45:00,0.0,0.0,24.873022,23.160919,0.0,22,0.0,0.0,2020-05-15,1
8,Anlage 1,2020-05-15 02:00:00,0.0,0.0,24.936930,23.026113,0.0,21,0.0,0.0,2020-05-15,2
9,Anlage 1,2020-05-15 02:15:00,0.0,0.0,25.012248,23.343229,0.0,21,0.0,0.0,2020-05-15,2


# Analyse:

## Analyse 1: Einfluss der Sonneneinstrahlung auf die Stromproduktion

### Hypothese

Je höher die Sonneneinstrahlung ist, desto höher ist die erzeugte AC-Leistung.

## Korrelation für Anlage 1 berechnen

In [67]:
# Daten von Anlage 1 auswählen

anlage1_zeitpunkt = zeitpunkt_df[zeitpunkt_df["anlage"] == "Anlage 1"]

In [68]:
# Zusammenhang zwischen Sonneneinstrahlung und AC-Leistung berechnen

korrelation_anlage1 = anlage1_zeitpunkt[
    ["sonneneinstrahlung", "ac_leistung"]
].corr()

korrelation_anlage1

,sonneneinstrahlung,ac_leistung
sonneneinstrahlung,1.000000,0.995864
ac_leistung,0.995864,1.000000


## Korrelation für Anlage 2 berechnen

In [69]:
# Daten von Anlage 2 auswählen

anlage2_zeitpunkt = zeitpunkt_df[zeitpunkt_df["anlage"] == "Anlage 2"]

In [70]:
# Zusammenhang zwischen Sonneneinstrahlung und AC-Leistung berechnen

korrelation_anlage2 = anlage2_zeitpunkt[["sonneneinstrahlung", "ac_leistung"]].corr()

korrelation_anlage2

,sonneneinstrahlung,ac_leistung
sonneneinstrahlung,1.000000,0.911084
ac_leistung,0.911084,1.000000


## Beide Werte übersichtlich ausgeben

In [71]:
# Korrelationswerte beider Anlagen anzeigen

print("Korrelation Anlage 1:",round(korrelation_anlage1.loc["sonneneinstrahlung", "ac_leistung"], 2))

print("Korrelation Anlage 2:",round(korrelation_anlage2.loc["sonneneinstrahlung", "ac_leistung"], 2))

Korrelation Anlage 1: 1.0
Korrelation Anlage 2: 0.91


## Streudiagramm erstellen

In [72]:
# Zusammenhang zwischen Sonneneinstrahlung und AC-Leistung darstellen

fig = px.scatter(
    zeitpunkt_df,
    x="sonneneinstrahlung",
    y="ac_leistung",
    color="anlage",
    color_discrete_map=FARBEN,
    opacity=0.5,
    title="Sonneneinstrahlung beeinflusst die AC-Leistung beider Anlagen stark",
    labels={
        "sonneneinstrahlung": "Sonneneinstrahlung",
        "ac_leistung": "AC-Leistung (kWh)",
        "anlage": "PV-Anlage"
    }
)

fig.show()

### Ergebnis

Bei beiden Anlagen besteht ein sehr starker positiver Zusammenhang zwischen Sonneneinstrahlung und AC-Leistung.

Die Korrelation beträgt bei Anlage 1 rund 1,00 und bei Anlage 2 rund 0,91. Mit steigender Sonneneinstrahlung steigt somit auch die erzeugte AC-Leistung.

Bei Anlage 2 ist jedoch eine stärkere Streuung zu erkennen. Mehrere Messpunkte zeigen trotz hoher Sonneneinstrahlung eine ungewöhnlich niedrige AC-Leistung. Dies kann auf zeitweise Leistungseinbrüche oder technische Auffälligkeiten hinweisen.

Die Hypothese wird bestätigt.

In [73]:
print("Korrelation Anlage 1:",round(korrelation_anlage1.loc["sonneneinstrahlung", "ac_leistung"], 3))

print("Korrelation Anlage 2:",round(korrelation_anlage2.loc["sonneneinstrahlung", "ac_leistung"], 3))

Korrelation Anlage 1: 0.996
Korrelation Anlage 2: 0.911


## Analyse 2: Temperatur und Stromproduktion

### Hypothese

Die Modultemperatur hängt stärker mit der AC-Leistung zusammen als die Außentemperatur.

## Korrelationen für Anlage 1 berechnen

In [74]:
# Korrelationen zwischen Wetter und AC-Leistung für Anlage 1

korrelation_temp_anlage1 = anlage1_zeitpunkt[["ac_leistung","aussentemperatur","modultemperatur","sonneneinstrahlung"]].corr().round(2)

korrelation_temp_anlage1

,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung
ac_leistung,1.00,0.73,0.96,1.00
aussentemperatur,0.73,1.00,0.85,0.72
modultemperatur,0.96,0.85,1.00,0.96
sonneneinstrahlung,1.00,0.72,0.96,1.00


## Korrelationen für Anlage 2 berechnen

In [75]:
# Korrelationen zwischen Wetter und AC-Leistung für Anlage 2

korrelation_temp_anlage2 = anlage2_zeitpunkt[[
    "ac_leistung",
    "aussentemperatur",
    "modultemperatur",
    "sonneneinstrahlung"
]].corr().round(2)

korrelation_temp_anlage2

,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung
ac_leistung,1.00,0.65,0.87,0.91
aussentemperatur,0.65,1.00,0.85,0.67
modultemperatur,0.87,0.85,1.00,0.95
sonneneinstrahlung,0.91,0.67,0.95,1.00


## Relevante Werte direkt vergleichen

In [76]:
# Temperatur-Korrelationen beider Anlagen übersichtlich ausgeben
print("Temperatur-Korrelationen:")
print("Anlage 1 - Außentemperatur:",korrelation_temp_anlage1.loc["ac_leistung", "aussentemperatur"])

print("Anlage 1 - Modultemperatur:",korrelation_temp_anlage1.loc["ac_leistung", "modultemperatur"])

print("Anlage 2 - Außentemperatur:",korrelation_temp_anlage2.loc["ac_leistung", "aussentemperatur"])

print("Anlage 2 - Modultemperatur:",korrelation_temp_anlage2.loc["ac_leistung", "modultemperatur"])

Temperatur-Korrelationen:
Anlage 1 - Außentemperatur: 0.73
Anlage 1 - Modultemperatur: 0.96
Anlage 2 - Außentemperatur: 0.65
Anlage 2 - Modultemperatur: 0.87


## Heatmap für Anlage 1

In [77]:
# Korrelationen von Anlage 1 als Heatmap darstellen

fig = px.imshow(
    korrelation_temp_anlage1,
    text_auto=True,
    aspect="auto",
    color_continuous_scale=["#1F9E96",  # Türkis: niedrige Werte
    "#F2F7F7",  # Fast Weiß: mittlere Werte
    "#0B3C5D"   # Dunkelblau: hohe Werte
    ],
    zmin=-1,
    zmax=1,
    title="Zusammenhänge zwischen Wetter und AC-Leistung - Anlage 1"
)

fig.show()

## Heatmap für Anlage 2

In [78]:
# Korrelationen von Anlage 2 als Heatmap darstellen

fig = px.imshow(
    korrelation_temp_anlage2,
    text_auto=True,
    aspect="auto",
    color_continuous_scale=["#1F9E96",  # Türkis: niedrige Werte
    "#F2F7F7",  # Fast Weiß: mittlere Werte
    "#0B3C5D"   # Dunkelblau: hohe Werte
    ],
    zmin=-1,
    zmax=1,
    title="Zusammenhänge zwischen Wetter und AC-Leistung - Anlage 2",
)

fig.show()

### Ergebnis

Bei beiden PV-Anlagen besteht zwischen der Modultemperatur und der AC-Leistung ein stärkerer Zusammenhang als zwischen der Außentemperatur und der AC-Leistung.

Bei Anlage 1 beträgt die Korrelation zwischen Modultemperatur und AC-Leistung 0,96. Die Korrelation mit der Außentemperatur liegt dagegen nur bei 0,73.

Bei Anlage 2 beträgt die Korrelation mit der Modultemperatur 0,87 und mit der Außentemperatur 0,65.

Die Hypothese wird daher bestätigt.

Die Ergebnisse bedeuten jedoch nicht, dass eine höhere Modultemperatur automatisch eine höhere Leistung verursacht. Die Module werden bei stärkerer Sonneneinstrahlung gleichzeitig wärmer und erzeugen mehr Strom. Die Sonneneinstrahlung ist deshalb weiterhin der wichtigste untersuchte Einflussfaktor.

## Zusatzprüfung: Modultemperatur bei ähnlicher Sonneneinstrahlung

Die bisherige Korrelation zeigt, dass Modultemperatur und AC-Leistung gemeinsam ansteigen. Daraus kann jedoch nicht geschlossen werden, dass eine höhere Modultemperatur die Leistung erhöht.

Starke Sonneneinstrahlung erwärmt die Module und erhöht gleichzeitig die Stromproduktion. Um den Temperatureffekt besser zu untersuchen, werden deshalb nur Zeitpunkte mit ähnlicher Sonneneinstrahlung zwischen 0,5 und 0,8 betrachtet.

Zusätzlich werden nur Zeitpunkte verwendet, an denen alle 22 Wechselrichter Messwerte geliefert haben.

## Vergleichsdaten erstellen

In [79]:
# Nur vollständige Zeitpunkte mit ähnlicher Einstrahlung auswählen

temperatur_pruefung_df = zeitpunkt_df.loc[
    (zeitpunkt_df["anzahl_wr"] == 22)
    & (zeitpunkt_df["sonneneinstrahlung"].between(0.5, 0.8))
].copy()

# Leistung je Wechselrichter und Einheit Einstrahlung berechnen

temperatur_pruefung_df["leistung_je_einstrahlung"] = (
    temperatur_pruefung_df["ac_je_wr"]
    / temperatur_pruefung_df["sonneneinstrahlung"]
)

# Modultemperaturen in Gruppen einteilen

temperatur_pruefung_df["temp_klasse"] = pd.cut(
    temperatur_pruefung_df["modultemperatur"],
    bins=[0, 40, 45, 50, 55, 80],
    labels=["unter 40", "40-45", "45-50", "50-55", "über 55"]
)

temperatur_pruefung_df.head()

,anlage,datum,dc_leistung,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung,anzahl_wr,ac_je_wr,ac_je_wr_relativ,tag,stunde,leistung_je_einstrahlung,temp_klasse
36,Anlage 1,2020-05-15 09:00:00,165598.892858,16211.350000,27.620970,45.009233,0.623153,22,736.879545,0.556131,2020-05-15,9,1182.502469,45-50
39,Anlage 1,2020-05-15 09:45:00,161370.035715,15793.112500,27.450108,40.930577,0.541206,22,717.868750,0.541784,2020-05-15,9,1326.424282,40-45
40,Anlage 1,2020-05-15 10:00:00,183026.214283,17896.798214,28.632192,52.547736,0.670675,22,813.490828,0.613951,2020-05-15,10,1212.942747,50-55
41,Anlage 1,2020-05-15 10:15:00,158716.553571,15531.950000,28.768913,47.633738,0.572283,22,705.997727,0.532825,2020-05-15,10,1233.650378,45-50
44,Anlage 1,2020-05-15 11:00:00,193768.035707,18936.559284,29.410819,47.685453,0.636561,22,860.752695,0.649620,2020-05-15,11,1352.192257,45-50


## Durchschnitt nach Temperaturgruppe berechnen

In [80]:
temperatur_vergleich_df = (
    temperatur_pruefung_df
    .groupby(
        ["anlage", "temp_klasse"],
        observed=True,
        as_index=False
    )
    .agg(
        durchschnitt_leistung_je_einstrahlung=(
            "leistung_je_einstrahlung", "mean"
        ),
        anzahl_messungen=(
            "leistung_je_einstrahlung", "count"
        )
    )
)

temperatur_vergleich_df[
    "durchschnitt_leistung_je_einstrahlung"
] = temperatur_vergleich_df[
    "durchschnitt_leistung_je_einstrahlung"
].round(0)

temperatur_vergleich_df

,anlage,temp_klasse,durchschnitt_leistung_je_einstrahlung,anzahl_messungen
0,Anlage 1,unter 40,1389.0,23
1,Anlage 1,40-45,1356.0,132
2,Anlage 1,45-50,1336.0,194
3,Anlage 1,50-55,1310.0,101
4,Anlage 1,über 55,1284.0,31
5,Anlage 2,unter 40,1240.0,34
6,Anlage 2,40-45,1046.0,65
7,Anlage 2,45-50,963.0,93
8,Anlage 2,50-55,930.0,71
9,Anlage 2,über 55,907.0,17


## Unterschied für Anlage 1 berechnen

In [81]:
anlage1_temp = temperatur_vergleich_df.loc[
    temperatur_vergleich_df["anlage"] == "Anlage 1"
].set_index("temp_klasse")

leistung_unter_40 = anlage1_temp.loc[
    "unter 40",
    "durchschnitt_leistung_je_einstrahlung"
]

leistung_ueber_55 = anlage1_temp.loc[
    "über 55",
    "durchschnitt_leistung_je_einstrahlung"
]

leistung_ueber_55 = float(leistung_ueber_55)
leistung_unter_40 = float(leistung_unter_40)

# Prozentualen Unterschied berechnen und auf eine Nachkommastelle runden
unterschied_prozent = round(
    (leistung_ueber_55 - leistung_unter_40)
    / leistung_unter_40 * 100,1)

print(
    "Veränderung bei über 55 °C gegenüber unter 40 °C:",
    unterschied_prozent,
    "%"
)

Veränderung bei über 55 °C gegenüber unter 40 °C: -7.6 %


In [82]:
fig = px.bar(
    temperatur_vergleich_df,
    x="temp_klasse",
    y="durchschnitt_leistung_je_einstrahlung",
    color="anlage",
    color_discrete_map=FARBEN,
    barmode="group",
    title=(
        "Bei ähnlicher Einstrahlung sinkt die Leistung "
        "bei höheren Modultemperaturen"
    ),
    labels={
        "temp_klasse": "Modultemperatur (°C)",
        "durchschnitt_leistung_je_einstrahlung":
            "Durchschnittliche Leistung je Einstrahlung",
        "anlage": "PV-Anlage"
    },
    text_auto=".0f"
)

fig.show()

## Analyse 3: Ertrag der einzelnen Wechselrichter

### Hypothese

Einzelne Wechselrichter erzeugen im untersuchten Zeitraum weniger Energie als die übrigen Wechselrichter derselben Anlage.

## Tagesertrag je Wechselrichter bestimmen

In [83]:
# Höchsten Tagesertrag jedes Wechselrichters 

wechselrichter_tag_df = analyse_df.groupby(
    ["anlage", "tag", "wechselrichter_id"],
    as_index=False
)["tagesertrag"].max()

wechselrichter_tag_df.head()

,anlage,tag,wechselrichter_id,tagesertrag
0,Anlage 1,2020-05-15,1BY6WEcLGh8j5v7,5754.0
1,Anlage 1,2020-05-15,1IF53ai7Xc0U56Y,6357.0
2,Anlage 1,2020-05-15,3PZuoBAID5Wc2HD,6274.0
3,Anlage 1,2020-05-15,7JYdWkrLSPkdwr4,6116.0
4,Anlage 1,2020-05-15,McdE0feGgRqW7Ca,6471.0


In [84]:
# Tageserträge für den gesamten Zeitraum 

wechselrichter_ertrag_df = wechselrichter_tag_df.groupby(["anlage", "wechselrichter_id"],as_index=False)["tagesertrag"].sum()

In [85]:
# Ergebnisspalte umbenennen

wechselrichter_ertrag_df = wechselrichter_ertrag_df.rename(columns={"tagesertrag": "ertrag_gesamt"})

wechselrichter_ertrag_df.head()

,anlage,wechselrichter_id,ertrag_gesamt
0,Anlage 1,1BY6WEcLGh8j5v7,225755.428571
1,Anlage 1,1IF53ai7Xc0U56Y,249859.416667
2,Anlage 1,3PZuoBAID5Wc2HD,249603.500000
3,Anlage 1,7JYdWkrLSPkdwr4,243859.000000
4,Anlage 1,McdE0feGgRqW7Ca,249620.464286


## Anzahl der Messzeilen und Tage berechnen

In [86]:
# Datenabdeckung je Wechselrichter berechnen

wechselrichter_daten_df = (
    analyse_df
    .groupby(
        ["anlage", "wechselrichter_id"],
        as_index=False
    )
    .agg(
        anzahl_messzeilen=("datum", "size"),
        anzahl_tage=("tag", "nunique")
    )
)

wechselrichter_daten_df.head()

,anlage,wechselrichter_id,anzahl_messzeilen,anzahl_tage
0,Anlage 1,1BY6WEcLGh8j5v7,3154,34
1,Anlage 1,1IF53ai7Xc0U56Y,3118,34
2,Anlage 1,3PZuoBAID5Wc2HD,3118,34
3,Anlage 1,7JYdWkrLSPkdwr4,3133,34
4,Anlage 1,McdE0feGgRqW7Ca,3124,34


## Informationen mit der Ertragstabelle verbinden

In [87]:
# Messzeilen und Tage zur Ertragstabelle hinzufügen

wechselrichter_ertrag_df = wechselrichter_ertrag_df.merge(
    wechselrichter_daten_df,
    on=["anlage", "wechselrichter_id"],
    how="left"
)

wechselrichter_ertrag_df.head()

,anlage,wechselrichter_id,ertrag_gesamt,anzahl_messzeilen,anzahl_tage
0,Anlage 1,1BY6WEcLGh8j5v7,225755.428571,3154,34
1,Anlage 1,1IF53ai7Xc0U56Y,249859.416667,3118,34
2,Anlage 1,3PZuoBAID5Wc2HD,249603.500000,3118,34
3,Anlage 1,7JYdWkrLSPkdwr4,243859.000000,3133,34
4,Anlage 1,McdE0feGgRqW7Ca,249620.464286,3124,34


## Durchschnitt je Anlage berechnen

In [88]:
# Durchschnittlichen Wechselrichter-Ertrag jeder Anlage ergänzen

wechselrichter_ertrag_df["durchschnitt_anlage"] = (wechselrichter_ertrag_df.groupby("anlage")["ertrag_gesamt"].transform("mean"))

wechselrichter_ertrag_df.head()

,anlage,wechselrichter_id,ertrag_gesamt,anzahl_messzeilen,anzahl_tage,durchschnitt_anlage
0,Anlage 1,1BY6WEcLGh8j5v7,225755.428571,3154,34,244581.76921
1,Anlage 1,1IF53ai7Xc0U56Y,249859.416667,3118,34,244581.76921
2,Anlage 1,3PZuoBAID5Wc2HD,249603.500000,3118,34,244581.76921
3,Anlage 1,7JYdWkrLSPkdwr4,243859.000000,3133,34,244581.76921
4,Anlage 1,McdE0feGgRqW7Ca,249620.464286,3124,34,244581.76921


## Abweichung vom Durchschnitt berechnen

In [89]:
# Prozentuale Abweichung vom Durchschnitt berechnen

wechselrichter_ertrag_df["abweichung_prozent"] = (
    (wechselrichter_ertrag_df["ertrag_gesamt"]- wechselrichter_ertrag_df["durchschnitt_anlage"])
    / wechselrichter_ertrag_df["durchschnitt_anlage"]* 100).round(2)


print("Beispiel zur Bedeutung:")
print("")
print("0 = genau im Durchschnitt")
print("")
print("-5 = 5 % unter dem Durchschnit" )
print("")
print("8 = 8 % über dem Durchschnitt")

wechselrichter_ertrag_df.head()

Beispiel zur Bedeutung:

0 = genau im Durchschnitt

-5 = 5 % unter dem Durchschnit

8 = 8 % über dem Durchschnitt


,anlage,wechselrichter_id,ertrag_gesamt,anzahl_messzeilen,anzahl_tage,durchschnitt_anlage,abweichung_prozent
0,Anlage 1,1BY6WEcLGh8j5v7,225755.428571,3154,34,244581.76921,-7.70
1,Anlage 1,1IF53ai7Xc0U56Y,249859.416667,3118,34,244581.76921,2.16
2,Anlage 1,3PZuoBAID5Wc2HD,249603.500000,3118,34,244581.76921,2.05
3,Anlage 1,7JYdWkrLSPkdwr4,243859.000000,3133,34,244581.76921,-0.30
4,Anlage 1,McdE0feGgRqW7Ca,249620.464286,3124,34,244581.76921,2.06


## Schwächste Wechselrichter anzeigen

In [90]:
# Wechselrichter nach Anlage und Ertrag sortieren

wechselrichter_ertrag_sortiert = wechselrichter_ertrag_df.sort_values(["anlage", "ertrag_gesamt"])

# Die drei schwächsten Wechselrichter jeder Anlage anzeigen

schwaechste_wechselrichter = wechselrichter_ertrag_sortiert.groupby("anlage").head(3)

schwaechste_wechselrichter.round({"ertrag_gesamt": 2,"durchschnitt_anlage": 2,"abweichung_prozent": 2})

,anlage,wechselrichter_id,ertrag_gesamt,anzahl_messzeilen,anzahl_tage,durchschnitt_anlage,abweichung_prozent
11,Anlage 1,bvBOhCH3iADSZry,222206.00,3155,34,244581.77,-9.15
0,Anlage 1,1BY6WEcLGh8j5v7,225755.43,3154,34,244581.77,-7.70
13,Anlage 1,ih0vzX44oOqAx2f,240939.40,3130,34,244581.77,-1.49
25,Anlage 2,Et9kgGMDl729KT4,155716.00,3195,34,203616.61,-23.52
33,Anlage 2,Quc1TzYxW2pYoWX,159300.80,3195,34,203616.61,-21.76
27,Anlage 2,LYwnQax7tkwH5Cb,164905.73,3259,34,203616.61,-19.01


## Datenabdeckung in Prozent berechnen

In [91]:
# Höchste Anzahl der Messzeilen innerhalb jeder Anlage bestimmen

wechselrichter_ertrag_df["max_messzeilen_anlage"] = (
    wechselrichter_ertrag_df
    .groupby("anlage")["anzahl_messzeilen"]
    .transform("max")
)

# Datenabdeckung jedes Wechselrichters berechnen

wechselrichter_ertrag_df["datenabdeckung_prozent"] = (
    wechselrichter_ertrag_df["anzahl_messzeilen"]
    / wechselrichter_ertrag_df["max_messzeilen_anlage"]
    * 100
).round(2)

## Beispiel-Wechselrichter kontrollieren


In [92]:
# Beispiel mit auffälliger Datenabdeckung prüfen

wechselrichter_ertrag_df.loc[
    wechselrichter_ertrag_df["wechselrichter_id"]
    == "mqwcsP2rE7J0TFp",
    [
        "anlage",
        "wechselrichter_id",
        "ertrag_gesamt",
        "abweichung_prozent",
        "anzahl_messzeilen",
        "anzahl_tage",
        "datenabdeckung_prozent"
    ]
]

,anlage,wechselrichter_id,ertrag_gesamt,abweichung_prozent,anzahl_messzeilen,anzahl_tage,datenabdeckung_prozent
36,Anlage 2,mqwcsP2rE7J0TFp,173033.0,-15.02,2355,26,72.26


## Auffällige Wechselrichter mit Datenabdeckung anzeigen

In [93]:
# Wechselrichter nach Ertragsabweichung sortieren

wechselrichter_kontrolle_df = (
    wechselrichter_ertrag_df[
        [
            "anlage",
            "wechselrichter_id",
            "ertrag_gesamt",
            "abweichung_prozent",
            "anzahl_messzeilen",
            "anzahl_tage",
            "datenabdeckung_prozent"
        ]
    ]
    .sort_values(
        ["anlage", "abweichung_prozent"]
    )
)

wechselrichter_kontrolle_df

,anlage,wechselrichter_id,ertrag_gesamt,abweichung_prozent,anzahl_messzeilen,anzahl_tage,datenabdeckung_prozent
11,Anlage 1,bvBOhCH3iADSZry,222206.000000,-9.15,3155,34,100.00
0,Anlage 1,1BY6WEcLGh8j5v7,225755.428571,-7.70,3154,34,99.97
13,Anlage 1,ih0vzX44oOqAx2f,240939.404762,-1.49,3130,34,99.21
19,Anlage 1,z9Y9gH1T5YWrNuG,243337.000000,-0.51,3125,34,99.05
9,Anlage 1,ZoEaEvLYb1n2sOq,243644.857143,-0.38,3123,34,98.99
3,Anlage 1,7JYdWkrLSPkdwr4,243859.000000,-0.30,3133,34,99.30
20,Anlage 1,zBIq5rxdHJRwDNY,243987.107143,-0.24,3119,34,98.86
6,Anlage 1,WRmjgnKYAwPKWDb,244791.750000,0.09,3118,34,98.83
16,Anlage 1,sjndEbLyjtCKgGv,244846.357143,0.11,3124,34,99.02
15,Anlage 1,rGa61gmuvPhdLxV,245401.000000,0.33,3124,34,99.02


## Spannweite der Abweichungen prüfen

In [94]:
# Kleinste und größte Abweichung je Anlage anzeigen

abweichung_spannweite_df = (
    wechselrichter_ertrag_df
    .groupby("anlage", as_index=False)
    .agg(
        kleinste_abweichung=(
            "abweichung_prozent", "min"
        ),
        groesste_abweichung=(
            "abweichung_prozent", "max"
        )
    )
)

abweichung_spannweite_df

,anlage,kleinste_abweichung,groesste_abweichung
0,Anlage 1,-9.15,3.48
1,Anlage 2,-23.52,28.64


### Datenqualität beim Wechselrichter-Ranking

Der Gesamtertrag eines Wechselrichters wurde aus den maximalen Tageserträgen des Untersuchungszeitraums berechnet. Wenn für einen Wechselrichter Messwerte oder ganze Tage fehlen, fällt sein berechneter Gesamtertrag automatisch niedriger aus.

Deshalb wurden zusätzlich die Anzahl der Messzeilen und die Anzahl der erfassten Tage je Wechselrichter untersucht.

Der Wechselrichter `mqwcsP2rE7J0TFp` liegt beispielsweise 15,02 % unter dem Durchschnitt, besitzt aber nur 2.355 Messzeilen. Andere Wechselrichter erreichen bis zu ungefähr 3.259 Messzeilen. Der niedrigere Gesamtertrag kann daher teilweise durch Datenlücken verursacht worden sein.

Bei Anlage 2 reicht die Abweichung vom Anlagendurchschnitt ungefähr von −23,52 % bis +28,6 %. Diese große Spannweite kann neben Datenlücken auch darauf hinweisen, dass die Wechselrichter beziehungsweise die angeschlossenen Modulgruppen unterschiedlich dimensioniert sind.

Da die installierte Leistung der Wechselrichter und Modulgruppen nicht bekannt ist, kann aus einem niedrigeren Gesamtertrag allein kein technischer Defekt abgeleitet werden.

## Balkendiagramm erstellen

In [95]:
# Gesamtertrag aller Wechselrichter vergleichen

fig = px.bar(
    wechselrichter_ertrag_df,
    x="wechselrichter_id",
    y="ertrag_gesamt",
    color="anlage",
    color_discrete_map=FARBEN,
    facet_col="anlage",
    title="Gesamtertrag der einzelnen Wechselrichter",
    labels={
        "wechselrichter_id": "Wechselrichter-ID",
        "ertrag_gesamt": "Gesamtertrag (kW)",
        "anlage": "PV-Anlage"
    }
)

fig.update_xaxes(showticklabels=False)

fig.show()

In [96]:
# Wechselrichter nach Anlage und Ertrag sortieren

wechselrichter_diagramm_df = wechselrichter_ertrag_df.sort_values(["anlage", "ertrag_gesamt"])

In [97]:
# Gesamtertrag der Wechselrichter übersichtlich darstellen

fig = px.bar(
    wechselrichter_diagramm_df,
    x="wechselrichter_id",
    y="ertrag_gesamt",
    color="anlage",
    color_discrete_map=FARBEN,
    facet_col="anlage",
    facet_col_spacing=0.08,
    title="Anlage 2 zeigt größere Ertragsunterschiede zwischen den Wechselrichtern",
    labels={
        "wechselrichter_id": "Wechselrichter-ID",
        "ertrag_gesamt": "Gesamtertrag (kW)",
        "anlage": "PV-Anlage"
    },
    hover_data={
        "wechselrichter_id": True,
        "ertrag_gesamt": ":.2f",
        "abweichung_prozent": ":.2f"
    }
)

# Für jede Anlage eine eigene X-Achse verwenden
fig.update_xaxes(
    matches=None,
    showticklabels=False
)

fig.show()

In [98]:
fig = px.bar(
    wechselrichter_diagramm_df,
    x="wechselrichter_id",
    y="ertrag_gesamt",
    color="anlage",
    color_discrete_map=FARBEN,
    barmode="overlay",
    opacity=0.55,
    title="Gesamtertrag der Wechselrichter: Anlage 1 und Anlage 2",
    labels={
        "wechselrichter_id": "Wechselrichter-ID",
        "ertrag_gesamt": "Gesamtertrag (kW)",
        "anlage": "PV-Anlage"
    },
    hover_data={
        "wechselrichter_id": True,
        "ertrag_gesamt": ":.2f",
        "abweichung_prozent": ":.2f"
    }
)

fig.show()

In [99]:


a1 = (wechselrichter_diagramm_df[
    wechselrichter_diagramm_df["anlage"] == "Anlage 1"
].sort_values("ertrag_gesamt").reset_index(drop=True))

a2 = (wechselrichter_diagramm_df[
    wechselrichter_diagramm_df["anlage"] == "Anlage 2"
].sort_values("ertrag_gesamt").reset_index(drop=True))

fig = go.Figure()

fig.add_bar(
    x=a1.index + 1,
    y=a1["ertrag_gesamt"],
    name="Anlage 1",
    marker_color="#1F9E96",
    customdata=a1["wechselrichter_id"],
    hovertemplate="Anlage 1<br>Wechselrichter-ID: %{customdata}<br>Gesamtertrag: %{y:,.0f}<extra></extra>"
)

fig.add_bar(
    x=a2.index + 1,
    y=a2["ertrag_gesamt"],
    name="Anlage 2",
    marker_color="#0B3C5D",
    customdata=a2["wechselrichter_id"],
    hovertemplate="Anlage 2<br>Wechselrichter-ID: %{customdata}<br>Gesamtertrag: %{y:,.0f}<extra></extra>"
)

fig.update_layout(
    barmode="overlay",
    title="Gesamtertrag der Wechselrichter beider Anlagen",
    xaxis_title="Wechselrichter, nach Ertrag sortiert",
    yaxis_title="Gesamtertrag (kW)"
)

fig.update_traces(opacity=0.6)
fig.show()

### Ergebnis

Die Hypothese wird bestätigt. Einzelne Wechselrichter erzeugten im untersuchten Zeitraum deutlich weniger Energie als der Durchschnitt ihrer jeweiligen Anlage.

Bei Anlage 1 fällt besonders der Wechselrichter `bvBOhCH3iADSZry` auf. Sein Gesamtertrag liegt 9,15 % unter dem Durchschnitt der Anlage. Der Wechselrichter `1BY6WEcLGh8j5v7` liegt 7,70 % darunter. Die übrigen Wechselrichter der Anlage weisen überwiegend ähnliche Erträge auf.

Bei Anlage 2 sind die Unterschiede deutlich größer. Der Wechselrichter `Et9kgGMDl729KT4` liegt 23,52 % unter dem Anlagendurchschnitt. Auch `Quc1TzYxW2pYoWX` mit −21,76 % und `LYwnQax7tkwH5Cb` mit −19,01 % zeigen auffällig niedrige Erträge.

Anlage 2 weist damit stärkere Leistungsunterschiede zwischen den Wechselrichtern auf. Die auffälligen Wechselrichter sollten genauer untersucht werden. Die Daten allein zeigen jedoch noch nicht, ob technische Defekte, Abschaltungen oder andere Ursachen dafür verantwortlich sind.

## Analyse 4: Auffällige Leistungseinbrüche

### Hypothese

Anlage 2 weist häufiger auffällige Leistungseinbrüche auf als Anlage 1.

Als auffällig gilt in dieser Analyse ein Messzeitpunkt, bei dem die Sonneneinstrahlung mindestens 60 % des Höchstwertes der jeweiligen Anlage erreicht, während die AC-Leistung unter 30 % ihres Höchstwertes liegt.

## Werte normalisieren

In [100]:
# AC-Leistung relativ zum Höchstwert jeder Anlage

zeitpunkt_df["ac_leistung_relativ"] = (zeitpunkt_df["ac_leistung"]/ zeitpunkt_df.groupby("anlage")["ac_leistung"].transform("max"))

zeitpunkt_df.head(3)

,anlage,datum,dc_leistung,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung,anzahl_wr,ac_je_wr,ac_je_wr_relativ,tag,stunde,ac_leistung_relativ
0,Anlage 1,2020-05-15 00:00:00,0.0,0.0,25.184316,22.857507,0.0,21,0.0,0.0,2020-05-15,0,0.0
1,Anlage 1,2020-05-15 00:15:00,0.0,0.0,25.084589,22.761668,0.0,21,0.0,0.0,2020-05-15,0,0.0
2,Anlage 1,2020-05-15 00:30:00,0.0,0.0,24.935753,22.592306,0.0,21,0.0,0.0,2020-05-15,0,0.0


In [101]:
# Sonneneinstrahlung relativ zum Höchstwert jeder Anlage berechnen

zeitpunkt_df["einstrahlung_relativ"] = (zeitpunkt_df["sonneneinstrahlung"] / zeitpunkt_df.groupby("anlage")["sonneneinstrahlung"].transform("max"))

zeitpunkt_df.head(3)

,anlage,datum,dc_leistung,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung,anzahl_wr,ac_je_wr,ac_je_wr_relativ,tag,stunde,ac_leistung_relativ,einstrahlung_relativ
0,Anlage 1,2020-05-15 00:00:00,0.0,0.0,25.184316,22.857507,0.0,21,0.0,0.0,2020-05-15,0,0.0,0.0
1,Anlage 1,2020-05-15 00:15:00,0.0,0.0,25.084589,22.761668,0.0,21,0.0,0.0,2020-05-15,0,0.0,0.0
2,Anlage 1,2020-05-15 00:30:00,0.0,0.0,24.935753,22.592306,0.0,21,0.0,0.0,2020-05-15,0,0.0,0.0


In [102]:
# Messungen mit hoher Sonneneinstrahlung auswählen
hohe_einstrahlung_korrigiert_df = zeitpunkt_df.loc[
    zeitpunkt_df["einstrahlung_relativ"] >= 0.60
].copy()

# Leistungseinbruch mit der Leistung je meldendem Wechselrichter prüfen
hohe_einstrahlung_korrigiert_df["leistungseinbruch_korrigiert"] = (
    hohe_einstrahlung_korrigiert_df["ac_je_wr_relativ"] < 0.30
)

# Ergebnisse je Anlage zusammenfassen
leistungseinbrueche_korrigiert_df = (
    hohe_einstrahlung_korrigiert_df
    .groupby("anlage", as_index=False)
    .agg(
        messungen_bei_hoher_einstrahlung=(
            "leistungseinbruch_korrigiert", "count"
        ),
        anzahl_leistungseinbrueche=(
            "leistungseinbruch_korrigiert", "sum"
        )
    )
)

# Prozentualen Anteil berechnen
leistungseinbrueche_korrigiert_df["anteil_prozent"] = (
    leistungseinbrueche_korrigiert_df["anzahl_leistungseinbrueche"]
    / leistungseinbrueche_korrigiert_df["messungen_bei_hoher_einstrahlung"]
    * 100
).round(2)

leistungseinbrueche_korrigiert_df

,anlage,messungen_bei_hoher_einstrahlung,anzahl_leistungseinbrueche,anteil_prozent
0,Anlage 1,293,0,0.00
1,Anlage 2,519,19,3.66


## Kontrollrechnung unter Berücksichtigung fehlender Wechselrichter-Meldungen

Bei der ursprünglichen Analyse wurde die AC-Leistung aller vorhandenen Wechselrichter je Zeitpunkt addiert. Wenn einzelne Wechselrichter keine Messwerte liefern, kann die gesamte AC-Leistung dadurch zu niedrig erscheinen.

Deshalb wird zusätzlich die durchschnittliche AC-Leistung je meldendem Wechselrichter verwendet. So lässt sich prüfen, wie stark fehlende Messwerte das ursprüngliche Ergebnis beeinflusst haben.

In [103]:
hohe_einstrahlung_korrigiert_df

,anlage,datum,dc_leistung,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung,anzahl_wr,ac_je_wr,ac_je_wr_relativ,tag,stunde,ac_leistung_relativ,einstrahlung_relativ,leistungseinbruch_korrigiert
49,Anlage 1,2020-05-15 12:15:00,209569.398819,20477.017856,31.507298,49.844493,0.787866,22,930.773539,0.702465,2020-05-15,12,0.702465,0.644919,False
51,Anlage 1,2020-05-15 12:45:00,209419.238093,20470.897620,32.391420,50.631240,0.761243,22,930.495346,0.702256,2020-05-15,12,0.702256,0.623126,False
56,Anlage 1,2020-05-15 14:00:00,222438.714294,21747.066071,33.763185,49.803904,0.735083,22,988.503003,0.746035,2020-05-15,14,0.746035,0.601713,False
57,Anlage 1,2020-05-15 14:15:00,259813.285720,25352.546429,34.130770,55.030613,0.893661,22,1152.388474,0.869721,2020-05-15,14,0.869721,0.731519,False
135,Anlage 1,2020-05-16 12:30:00,228677.553569,22351.071428,30.886469,50.449570,0.767119,22,1015.957792,0.766755,2020-05-16,12,0.766755,0.627936,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6169,Anlage 2,2020-06-15 10:15:00,12550.120000,12251.326667,30.603249,47.439764,0.766245,22,556.878485,0.471572,2020-06-15,10,0.471572,0.697369,False
6173,Anlage 2,2020-06-15 11:15:00,12056.030000,11779.477619,30.406593,40.686538,0.663464,22,535.430801,0.453410,2020-06-15,11,0.453410,0.603826,False
6177,Anlage 2,2020-06-15 12:15:00,12718.292381,12413.529048,31.543530,48.458500,0.820264,22,564.251320,0.477815,2020-06-15,12,0.477815,0.746532,False
6184,Anlage 2,2020-06-15 14:00:00,14503.103810,14166.742381,28.538208,41.651166,0.704287,22,643.942835,0.545299,2020-06-15,14,0.545299,0.640980,False


In [104]:
leistungseinbrueche_korrigiert_df

,anlage,messungen_bei_hoher_einstrahlung,anzahl_leistungseinbrueche,anteil_prozent
0,Anlage 1,293,0,0.00
1,Anlage 2,519,19,3.66


Die ursprüngliche Berechnung erkannte bei Anlage 2 insgesamt 30 auffällige Messzeitpunkte. Nach Berücksichtigung der Anzahl meldender Wechselrichter blieben 19 Auffälligkeiten bestehen. Datenlücken beeinflussten somit die genaue Anzahl, die grundsätzliche Aussage blieb jedoch gleich: Anlage 2 weist häufiger Auffälligkeiten auf als Anlage 1.

In [105]:
# kontrolle
zeitpunkt_df[["anlage","ac_leistung","ac_leistung_relativ","sonneneinstrahlung","einstrahlung_relativ"]].head(3)

,anlage,ac_leistung,ac_leistung_relativ,sonneneinstrahlung,einstrahlung_relativ
0,Anlage 1,0.0,0.0,0.0,0.0
1,Anlage 1,0.0,0.0,0.0,0.0
2,Anlage 1,0.0,0.0,0.0,0.0


In [106]:
# Kontrolle der zusätzlichen Werte je meldendem Wechselrichter
zeitpunkt_df[[
    "anlage",
    "datum",
    "anzahl_wr",
    "ac_leistung",
    "ac_je_wr",
    "ac_je_wr_relativ",
    "einstrahlung_relativ"
]].head(3)

,anlage,datum,anzahl_wr,ac_leistung,ac_je_wr,ac_je_wr_relativ,einstrahlung_relativ
0,Anlage 1,2020-05-15 00:00:00,21,0.0,0.0,0.0,0.0
1,Anlage 1,2020-05-15 00:15:00,21,0.0,0.0,0.0,0.0
2,Anlage 1,2020-05-15 00:30:00,21,0.0,0.0,0.0,0.0


## Kontrollrechnung unter Berücksichtigung fehlender Wechselrichter-Meldungen

Bei der ursprünglichen Analyse wurde die gesamte AC-Leistung aller vorhandenen Wechselrichter je Zeitpunkt verwendet. Wenn einzelne Wechselrichter keine Messwerte liefern, kann die Gesamtleistung dadurch zu niedrig erscheinen.

Deshalb wird zusätzlich die durchschnittliche AC-Leistung je meldendem Wechselrichter untersucht.

## Zeitpunkte mit hoher Einstrahlung auswählen


In [107]:
# Messungen mit hoher Sonneneinstrahlung auswählen
hohe_einstrahlung_korrigiert_df = zeitpunkt_df.loc[
    zeitpunkt_df["einstrahlung_relativ"] >= 0.60
].copy()

# Korrigierte Leistungseinbrüche markieren
hohe_einstrahlung_korrigiert_df["leistungseinbruch_korrigiert"] = (
    hohe_einstrahlung_korrigiert_df["ac_je_wr_relativ"] < 0.30
)

# Ergebnisse je Anlage zusammenfassen
leistungseinbrueche_korrigiert_df = (
    hohe_einstrahlung_korrigiert_df
    .groupby("anlage", as_index=False)
    .agg(
        messungen_bei_hoher_einstrahlung=(
            "leistungseinbruch_korrigiert", "count"
        ),
        anzahl_leistungseinbrueche=(
            "leistungseinbruch_korrigiert", "sum"
        )
    )
)

# Prozentualen Anteil berechnen
leistungseinbrueche_korrigiert_df["anteil_prozent"] = (
    leistungseinbrueche_korrigiert_df["anzahl_leistungseinbrueche"]
    / leistungseinbrueche_korrigiert_df["messungen_bei_hoher_einstrahlung"]
    * 100
).round(2)

leistungseinbrueche_korrigiert_df

,anlage,messungen_bei_hoher_einstrahlung,anzahl_leistungseinbrueche,anteil_prozent
0,Anlage 1,293,0,0.00
1,Anlage 2,519,19,3.66


### Ergebnis der Kontrollrechnung

Bei der ursprünglichen Berechnung wurden die AC-Leistungen aller vorhandenen Wechselrichter je Zeitpunkt addiert. Da bei Anlage 2 häufiger Messwerte einzelner Wechselrichter fehlen, konnte die Gesamtleistung dadurch zu niedrig erscheinen.

Deshalb wurde zusätzlich die durchschnittliche AC-Leistung je meldendem Wechselrichter verwendet.

Bei Anlage 1 wurden weiterhin keine auffälligen Leistungseinbrüche festgestellt.

Bei Anlage 2 sank die Anzahl der auffälligen Messzeitpunkte von **30 auf 19**. Der Anteil sank dadurch von **5,78 % auf 3,66 %**.

Die fehlenden Wechselrichter-Meldungen haben die ursprüngliche Anzahl der Auffälligkeiten somit teilweise beeinflusst. Die Kernaussage der Analyse bleibt jedoch bestehen: Auch nach Berücksichtigung der Datenlücken treten die auffälligen Leistungseinbrüche ausschließlich bei Anlage 2 auf.

In [108]:
# Nur Messungen mit mindestens 60 % der maximalen Einstrahlung auswählen

hohe_einstrahlung_df = zeitpunkt_df[zeitpunkt_df["einstrahlung_relativ"] >= 0.60].copy()

hohe_einstrahlung_df.head()


,anlage,datum,dc_leistung,ac_leistung,aussentemperatur,modultemperatur,sonneneinstrahlung,anzahl_wr,ac_je_wr,ac_je_wr_relativ,tag,stunde,ac_leistung_relativ,einstrahlung_relativ
49,Anlage 1,2020-05-15 12:15:00,209569.398819,20477.017856,31.507298,49.844493,0.787866,22,930.773539,0.702465,2020-05-15,12,0.702465,0.644919
51,Anlage 1,2020-05-15 12:45:00,209419.238093,20470.897620,32.391420,50.631240,0.761243,22,930.495346,0.702256,2020-05-15,12,0.702256,0.623126
56,Anlage 1,2020-05-15 14:00:00,222438.714294,21747.066071,33.763185,49.803904,0.735083,22,988.503003,0.746035,2020-05-15,14,0.746035,0.601713
57,Anlage 1,2020-05-15 14:15:00,259813.285720,25352.546429,34.130770,55.030613,0.893661,22,1152.388474,0.869721,2020-05-15,14,0.869721,0.731519
135,Anlage 1,2020-05-16 12:30:00,228677.553569,22351.071428,30.886469,50.449570,0.767119,22,1015.957792,0.766755,2020-05-16,12,0.766755,0.627936


## Leistungseinbrüche markieren

In [109]:
# Leistung unter 30 % trotz hoher Einstrahlung als auffällig markieren

hohe_einstrahlung_df["leistungseinbruch"] = (hohe_einstrahlung_df["ac_leistung_relativ"] < 0.30)

In [110]:
# kontrolle

hohe_einstrahlung_df[["anlage","datum","einstrahlung_relativ","ac_leistung_relativ","leistungseinbruch"]].head()

,anlage,datum,einstrahlung_relativ,ac_leistung_relativ,leistungseinbruch
49,Anlage 1,2020-05-15 12:15:00,0.644919,0.702465,False
51,Anlage 1,2020-05-15 12:45:00,0.623126,0.702256,False
56,Anlage 1,2020-05-15 14:00:00,0.601713,0.746035,False
57,Anlage 1,2020-05-15 14:15:00,0.731519,0.869721,False
135,Anlage 1,2020-05-16 12:30:00,0.627936,0.766755,False


## Auffällige Messungen bei hoher Sonneneinstrahlung prüfen

In [111]:
# Messungen mit niedriger relativer AC-Leistung anzeigen
auffaellige_werte = hohe_einstrahlung_df.loc[
    hohe_einstrahlung_df["leistungseinbruch"],
    ["anlage", "datum", "sonneneinstrahlung", "ac_leistung", "ac_leistung_relativ"]
].sort_values(["anlage", "datum"])

auffaellige_werte

,anlage,datum,sonneneinstrahlung,ac_leistung,ac_leistung_relativ
3581,Anlage 2,2020-05-19 10:15:00,0.775781,3355.366190,0.129153
3675,Anlage 2,2020-05-20 10:00:00,0.747510,7785.319524,0.299669
3676,Anlage 2,2020-05-20 10:15:00,0.864114,5346.840000,0.205808
3677,Anlage 2,2020-05-20 10:30:00,0.849222,3496.646667,0.134591
3678,Anlage 2,2020-05-20 10:45:00,0.912170,3669.426667,0.141242
3679,Anlage 2,2020-05-20 11:00:00,0.878214,3557.437143,0.136931
3680,Anlage 2,2020-05-20 11:15:00,0.900284,3680.076190,0.141652
3681,Anlage 2,2020-05-20 11:30:00,0.904426,3701.473333,0.142475
3682,Anlage 2,2020-05-20 11:45:00,0.892505,3843.013333,0.147923
3683,Anlage 2,2020-05-20 12:00:00,0.888746,5127.686667,0.197372


In [112]:
FARBEN = {
    "einstrahlung_relativ": "#1F9E96",
    "ac_leistung_relativ": "#0B3C5D"
}

In [113]:
# Anlage 2 am 20.05.2020 im Tagesverlauf untersuchen
tag_df = zeitpunkt_df.loc[
    (zeitpunkt_df["anlage"] == "Anlage 2")
    & (zeitpunkt_df["datum"].dt.date == pd.Timestamp("2020-05-20").date())
].sort_values("datum")

fig = px.line(
    tag_df,
    x="datum",
    y=["einstrahlung_relativ", "ac_leistung_relativ"],
    color_discrete_map=FARBEN,
    title="Anlage 2 am 20.05.2020: Einstrahlung und AC-Leistung",
    labels={"datum": "Uhrzeit", "value": "Relativer Wert", "variable": "Messgröße"},
)

fig.update_xaxes(tickformat="%H:%M")

fig.show()

### Einstrahlung und AC-Leistung am 20.05.2020

Das Diagramm vergleicht für **Anlage 2** die Sonneneinstrahlung und die AC-Leistung im Tagesverlauf. Beide Werte sind relativ zum jeweils höchsten gemessenen Wert der Anlage dargestellt. **1,0 entspricht also 100 % des jeweiligen Höchstwerts.**

Zwischen **10:00 und 14:00 Uhr** liegen 17 aufeinanderfolgende Messungen vor, bei denen die Sonneneinstrahlung hoch ist, die AC-Leistung aber unter 30 % ihres Höchstwerts bleibt. Dieser Zeitraum ist auffällig und sollte genauer untersucht werden. Aus dem Diagramm allein lässt sich die Ursache der niedrigen Leistung nicht bestimmen.

In [114]:
# Beide Anlagen am 20.05.2020 vergleichen
vergleich_df = zeitpunkt_df.loc[
    zeitpunkt_df["datum"].dt.date == pd.Timestamp("2020-05-20").date()
].sort_values("datum")

fig = px.line(
    vergleich_df,
    x="datum",
    y="ac_leistung_relativ",
    color="anlage",
    color_discrete_map=FARBEN,
    title="Anlage 2 zeigt am 20.05.2020 einen deutlichen Leistungseinbruch",
    labels={
        "datum": "Uhrzeit",
        "ac_leistung_relativ": "AC-Leistung relativ zum Anlagenhöchstwert",
        "anlage": "PV-Anlage"
    }
)

# Auf der X-Achse nur die Uhrzeit anzeigen

fig.update_xaxes(tickformat="%H:%M")

fig.show()

### Vergleich der AC-Leistung beider Anlagen am 20.05.2020

Das Diagramm vergleicht den Tagesverlauf der relativen AC-Leistung beider PV-Anlagen. Die Leistung jeder Anlage wurde durch ihren eigenen Höchstwert geteilt. Dadurch können die Kurven trotz unterschiedlicher Leistungsgrößen miteinander verglichen werden.

Anlage 1 zeigt einen gleichmäßigen Tagesverlauf: Die Leistung steigt am Vormittag an, erreicht ungefähr zur Mittagszeit ihren Höchstwert und sinkt anschließend wieder.

Bei Anlage 2 steigt die Leistung zunächst ebenfalls an. Zwischen 10:00 und 14:00 Uhr fällt sie jedoch deutlich ab und bleibt überwiegend unter 30 % ihres Höchstwertes. Gegen 14:15 Uhr steigt sie plötzlich wieder stark an.

Der Unterschied zwischen den Anlagen unterstützt die Vermutung, dass bei Anlage 2 ein anlagenspezifischer Leistungseinbruch vorlag. Der Vergleich der AC-Leistung allein reicht jedoch nicht als Beweis aus. Erst zusammen mit dem vorherigen Diagramm, das für Anlage 2 gleichzeitig eine hohe Sonneneinstrahlung zeigt, wird die Auffälligkeit deutlich.

Die genaue Ursache kann mit den vorhandenen Daten nicht festgestellt werden.

In [115]:
FARBEN = {
    "Anlage 1": "#1F9E96",
    "Anlage 2": "#0B3C5D"
}

In [116]:
# Beide Anlagen am 20.05.2020 vergleichen

# Vollständigen Zeitraster für den 20.05.2020 erstellen
zeitraster = pd.date_range(
    start="2020-05-20 00:00:00",
    periods=96,
    freq="15min"
)

vergleich_liste = []

# Fehlende Zeitpunkte für jede Anlage sichtbar machen
for anlage in ["Anlage 1", "Anlage 2"]:

    anlage_tag = (
        zeitpunkt_df.loc[
            (zeitpunkt_df["anlage"] == anlage)
            & (
                zeitpunkt_df["datum"].dt.date
                == pd.Timestamp("2020-05-20").date()
            )
        ]
        .set_index("datum")
        .reindex(zeitraster)
    )

    anlage_tag["anlage"] = anlage
    anlage_tag.index.name = "datum"

    vergleich_liste.append(anlage_tag.reset_index())

# Beide Anlagen wieder zusammenfügen
vergleich_df = pd.concat(
    vergleich_liste,
    ignore_index=True
)

# Diagramm erstellen
fig = px.line(
    vergleich_df,
    x="datum",
    y="ac_leistung_relativ",
    color="anlage",
    color_discrete_map=FARBEN,
    title="Anlage 2 zeigt am 20.05.2020 einen deutlichen Leistungseinbruch",
    labels={
        "datum": "Uhrzeit",
        "ac_leistung_relativ": "Relative AC-Leistung",
        "anlage": "PV-Anlage"
    }
)

# Fehlende Werte nicht mit Linien verbinden
fig.update_traces(connectgaps=False)

# Nur Uhrzeit anzeigen
fig.update_xaxes(tickformat="%H:%M")

fig.show()

### Vergleich der AC-Leistung beider Anlagen am 20.05.2020

Das Diagramm vergleicht die relative AC-Leistung beider PV-Anlagen am 20.05.2020.

Anlage 1 zeigt bis 13:15 Uhr einen gleichmäßigen Anstieg der Leistung. Zwischen 13:30 und 17:15 Uhr fehlen jedoch 16 aufeinanderfolgende Messzeitpunkte. Deshalb wird die Linie in diesem Zeitraum unterbrochen. Ab 17:30 Uhr sind wieder Messwerte vorhanden.

Bei Anlage 2 sind für den gesamten Tag Messwerte vorhanden. Zwischen 10:00 und 14:00 Uhr bleibt die relative AC-Leistung trotz hoher Sonneneinstrahlung unter 30 % ihres Höchstwertes.

Aufgrund der Datenlücke kann der vollständige Tagesverlauf von Anlage 1 an diesem Tag nicht beurteilt werden. Die genaue Ursache des Leistungseinbruchs bei Anlage 2 kann mit den vorhandenen Daten ebenfalls nicht bestimmt werden.

In [117]:
# Anlage 2 am 20.05.2020 im Tagesverlauf untersuchen

tag_df = zeitpunkt_df.loc[
    (zeitpunkt_df["anlage"] == "Anlage 2")
    & (zeitpunkt_df["datum"].dt.date == pd.Timestamp("2020-05-20").date())
].sort_values("datum")

In [118]:
# Relative Sonneneinstrahlung und AC-Leistung vergleichen

fig = px.line(
    tag_df,
    x="datum",
    y=["einstrahlung_relativ", "ac_leistung_relativ"],
    title="Deutlicher Leistungseinbruch bei Anlage 2 am 20.05.2020",
    labels={
        "datum": "Uhrzeit",
        "value": "Relativer Wert",
        "variable": "Messgröße"
    },
    color_discrete_map={
        "einstrahlung_relativ": "#1F9E96",
        "ac_leistung_relativ": "#0B3C5D"
    }
)

# Uhrzeit ohne vollständiges Datum anzeigen
fig.update_xaxes(tickformat="%H:%M")

# Namen in der Legende auf Deutsch anzeigen
neue_namen = {
    "einstrahlung_relativ": "Sonneneinstrahlung",
    "ac_leistung_relativ": "AC-Leistung"
}

fig.for_each_trace(
    lambda linie: linie.update(name=neue_namen[linie.name])
)

fig.show()

### Leistungseinbruch am 20.05.2020

Der Tagesverlauf von Anlage 2 zeigt am 20.05.2020 einen besonders deutlichen Leistungseinbruch.

Zwischen ungefähr 10:15 und 14:00 Uhr bleibt die relative Sonneneinstrahlung auf einem hohen Niveau. Die relative AC-Leistung fällt während dieses Zeitraums jedoch deutlich ab und liegt größtenteils nur zwischen etwa 13 % und 21 % ihres Höchstwertes.

Gegen 14:15 Uhr steigt die AC-Leistung plötzlich wieder an, obwohl die Sonneneinstrahlung zu diesem Zeitpunkt bereits abnimmt. Der mehrstündige Verlauf deutet deshalb auf eine vorübergehende Störung oder Abschaltung hin.

Die genaue Ursache kann anhand der vorhandenen Daten nicht festgestellt werden. Das Diagramm bestätigt jedoch, dass bei Anlage 2 ein längerer und deutlich erkennbarer Leistungseinbruch vorlag.

## Zusatzprüfung: Wechselrichter von Anlage 2 am 20.05.2020

Um den Leistungseinbruch genauer zu untersuchen, werden die einzelnen Wechselrichter von Anlage 2 zwischen 10:00 und 14:00 Uhr betrachtet.

Dabei wird geprüft:

- wie viele der 22 Wechselrichter Messwerte geliefert haben,
- wie viele im Durchschnitt keine AC-Leistung erzeugten,
- wie viele eine sehr niedrige durchschnittliche AC-Leistung aufwiesen,
- und wie viele mehr als 1.000 kW erreichten.

In [119]:
# Anlage 2 am 20.05.2020 zwischen 10:00 und 14:00 Uhr auswählen

fenster = analyse_df.loc[
    (analyse_df["anlage"] == "Anlage 2")
    & (analyse_df["datum"] >= "2020-05-20 10:00")
    & (analyse_df["datum"] <= "2020-05-20 14:00")
].copy()

# Durchschnittliche AC-Leistung je Wechselrichter berechnen

wr = (
    fenster
    .groupby("wechselrichter_id")["ac_leistung"]
    .mean()
    .sort_values()
)

# Ergebnisse ausgeben

print("Meldende Wechselrichter:", len(wr), "von 22")
print("Keine Messwerte:", 22 - len(wr))
print("Exakt 0 kW:", (wr == 0).sum())

print(
    "Unter 200 kW, aber über 0:",
    ((wr > 0) & (wr < 200)).sum()
)

print("Über 1.000 kW:", (wr > 1000).sum())

wr

Meldende Wechselrichter: 12 von 22
Keine Messwerte: 10
Exakt 0 kW: 3
Unter 200 kW, aber über 0: 5
Über 1.000 kW: 3


wechselrichter_id
oZZkBaNadn6DNKz       0.000000
PeE6FRyGXUgsRhN       0.000000
xoJJ8DcxJEcupym       0.000000
81aHJ1q11NBPMrL      35.863866
q49J1IKaHRwDQnt      55.992437
LYwnQax7tkwH5Cb      81.804314
LlT2YUhhzqhg5Sw     102.358431
9kRcWv60rDACzjR     119.386947
rrq4fwE8jgrTyWY     661.253277
V94E5Ben1TlhnDV    1195.156975
vOuJvMaM2sgwLmb    1210.522745
WcxssY2VbP4hApt    1223.138964
Name: ac_leistung, dtype: float64

In [120]:
# Wechselrichter zwischen 200 und 1.000 kW anzeigen

wr.loc[(wr >= 200) & (wr <= 1000)]

wechselrichter_id
rrq4fwE8jgrTyWY    661.253277
Name: ac_leistung, dtype: float64

In [121]:
# Einheitliche Farben für alle Diagramme

FARBEN = {
    "wechselrichter_id": "#1F9E96",
    "durchschnittliche_ac_leistung": "#0B3C5D"
}

In [122]:
# Ergebnisse als Diagramm

wr_df = wr.reset_index(name="durchschnittliche_ac_leistung")

fig = px.bar(
    wr_df,
    x="wechselrichter_id",
    y="durchschnittliche_ac_leistung",
    color_discrete_sequence=[FARBEN["durchschnittliche_ac_leistung"]],
    title=(
        "Anlage 2 am 20.05.2020, 10-14 Uhr: "
        "Nur 12 von 22 Wechselrichtern liefern Messwerte"
    ),
    labels={
        "wechselrichter_id": "Wechselrichter-ID",
        "durchschnittliche_ac_leistung": "Durchschnittliche AC-Leistung"
    }
)

fig.update_xaxes(showticklabels=False)

fig.show()

### Ergebnis der Wechselrichter-Prüfung

Im untersuchten Zeitraum lieferten nur 12 von insgesamt 22 Wechselrichtern Messwerte. Für zehn Wechselrichter liegen zwischen 10:00 und 14:00 Uhr keine Daten vor.

Von den zwölf meldenden Wechselrichtern erzeugten drei im Durchschnitt keine AC-Leistung. Weitere fünf lagen über 0, aber unter 200. Drei Wechselrichter erreichten mehr als 1.000. Ein weiterer Wechselrichter lag mit ungefähr 661 zwischen diesen Gruppen.

Die fehlenden Meldungen erklären einen Teil der niedrigen Gesamtleistung von Anlage 2. Gleichzeitig zeigen mehrere der vorhandenen Messwerte eine ungewöhnlich niedrige AC-Leistung.

Die drei Wechselrichter mit durchschnittlich 0 sind nicht dieselben Wechselrichter, die in der Gesamtertragsanalyse die stärksten negativen Abweichungen zeigten. Der kurzfristige Leistungseinbruch am 20.05.2020 und die langfristig niedrigen Gesamterträge einzelner Wechselrichter sind daher zwei unterschiedliche Auffälligkeiten.

In [123]:
# Zusammenhang zwischen Sonneneinstrahlung und AC-Leistung

fig = px.scatter(
    zeitpunkt_df,
    x="einstrahlung_relativ",
    y="ac_leistung_relativ",
    color="anlage",
    color_discrete_map={
        "Anlage 1": "#1F9E96",
        "Anlage 2": "#0B3C5D"
    },
    opacity=0.7,
    title="Zusammenhang zwischen Sonneneinstrahlung und AC-Leistung",
    labels={
        "einstrahlung_relativ": "Relative Sonneneinstrahlung",
        "ac_leistung_relativ": "Relative AC-Leistung",
        "anlage": "PV-Anlage"
    }
)

fig.update_traces(marker_size=5)
fig.show()

In [124]:
zeitpunkt_df.groupby("anlage").apply(
    lambda df: df["einstrahlung_relativ"].corr(df["ac_leistung_relativ"]),
    include_groups=False
).round(2)

anlage
Anlage 1    1.00
Anlage 2    0.91
dtype: float64

### Mehr Sonneneinstrahlung bedeutet mehr AC-Leistung

Das Streudiagramm zeigt den Zusammenhang zwischen der relativen Sonneneinstrahlung und der relativen AC-Leistung beider PV-Anlagen. Jeder Punkt steht für einen Messzeitpunkt.

Bei **Anlage 1** liegen die Punkte nahezu auf einer Linie: Steigt die Sonneneinstrahlung, steigt auch die AC-Leistung. Bei **Anlage 2** ist ebenfalls ein positiver Zusammenhang zu erkennen, die Werte streuen jedoch stärker. Besonders bei hoher Einstrahlung gibt es Messungen mit vergleichsweise niedriger Leistung.

Die Korrelationswerte betragen **r = 1,00** für Anlage 1 und **r = 0,91** für Anlage 2. Damit wird die Hypothese grundsätzlich bestätigt. Die Abweichungen bei Anlage 2 untersuchen wir im nächsten Schritt genauer.

## Anzahl und Anteil berechnen

In [125]:
# Leistungseinbrüche je Anlage zusammenfassen

leistungseinbrueche_df = hohe_einstrahlung_df.groupby("anlage",as_index=False).agg(messungen_bei_hoher_einstrahlung=("leistungseinbruch", "count"),anzahl_leistungseinbrueche=("leistungseinbruch", "sum"))

In [126]:
# Prozentualen Anteil der Leistungseinbrüche berechnen

leistungseinbrueche_df["anteil_prozent"] = (leistungseinbrueche_df["anzahl_leistungseinbrueche"]/ leistungseinbrueche_df["messungen_bei_hoher_einstrahlung"]* 100).round(2)

leistungseinbrueche_df

,anlage,messungen_bei_hoher_einstrahlung,anzahl_leistungseinbrueche,anteil_prozent
0,Anlage 1,293,0,0.00
1,Anlage 2,519,30,5.78


In [127]:
print(hohe_einstrahlung_df.columns.tolist())
print(hohe_einstrahlung_df.groupby("anlage")["leistungseinbruch"].value_counts(dropna=False))

['anlage', 'datum', 'dc_leistung', 'ac_leistung', 'aussentemperatur', 'modultemperatur', 'sonneneinstrahlung', 'anzahl_wr', 'ac_je_wr', 'ac_je_wr_relativ', 'tag', 'stunde', 'ac_leistung_relativ', 'einstrahlung_relativ', 'leistungseinbruch']
anlage    leistungseinbruch
Anlage 1  False                293
Anlage 2  False                489
          True                  30
Name: count, dtype: int64


In [128]:
hohe_einstrahlung_df[["anlage", "sonneneinstrahlung", "ac_leistung", "ac_leistung_relativ", "leistungseinbruch"]].loc[hohe_einstrahlung_df["leistungseinbruch"]].head(10)

,anlage,sonneneinstrahlung,ac_leistung,ac_leistung_relativ,leistungseinbruch
3581,Anlage 2,0.775781,3355.366190,0.129153,True
3675,Anlage 2,0.747510,7785.319524,0.299669,True
3676,Anlage 2,0.864114,5346.840000,0.205808,True
3677,Anlage 2,0.849222,3496.646667,0.134591,True
3678,Anlage 2,0.912170,3669.426667,0.141242,True
3679,Anlage 2,0.878214,3557.437143,0.136931,True
3680,Anlage 2,0.900284,3680.076190,0.141652,True
3681,Anlage 2,0.904426,3701.473333,0.142475,True
3682,Anlage 2,0.892505,3843.013333,0.147923,True
3683,Anlage 2,0.888746,5127.686667,0.197372,True


## Ergebnis als Balkendiagramm darstellen

In [129]:
FARBEN = {
    "Anlage 1": "#1F9E96",
    "Anlage 2": "#0B3C5D"
}

In [130]:
# Anteil auffälliger Leistungseinbrüche vergleichen

fig = px.bar(
    leistungseinbrueche_df,
    x="anlage",
    y="anteil_prozent",
    color="anlage",
    color_discrete_map=FARBEN,
    text="anteil_prozent",
    title="Auffällige Leistungseinbrüche treten ausschließlich bei Anlage 2 auf",
    labels={
        "anlage": "PV-Anlage",
        "anteil_prozent": "Anteil der Leistungseinbrüche in %"
    }
)

fig.update_traces(texttemplate="%{text:.2f} %", textposition="outside")

fig.show()

In [131]:
fig = px.bar(
    leistungseinbrueche_df,
    x="anteil_prozent",
    y="anlage",
    orientation="h",
    text="anteil_prozent",
    title="Leistungseinbrüche bei hoher Sonneneinstrahlung",
    labels={
        "anlage": "PV-Anlage",
        "anteil_prozent": "Anteil der Messungen mit Leistungseinbruch (%)"
    },
    color_discrete_sequence=["#0B3C5D"]
)

fig.update_traces(texttemplate="%{text:.2f} %", textposition="outside")
fig.update_xaxes(range=[0, 7])
fig.update_layout(showlegend=False, height=350)
fig.show()

### Ergebnis

Die Hypothese wird bestätigt.

Bei Anlage 1 wurden 293 Messzeitpunkte mit hoher Sonneneinstrahlung untersucht. Dabei wurde kein auffälliger Leistungseinbruch festgestellt. Der Anteil beträgt somit 0,00 %.

Bei Anlage 2 wurden 519 Messzeitpunkte mit hoher Sonneneinstrahlung untersucht. Bei 30 dieser Messungen lag die relative AC-Leistung trotz hoher Sonneneinstrahlung unter 30 %. Das entspricht einem Anteil von 5,78 %.

Auffällige Leistungseinbrüche wurden nach der verwendeten Analyseregel somit ausschließlich bei Anlage 2 erkannt. Dieses Ergebnis passt auch zur stärkeren Streuung im vorherigen Streudiagramm.

Die Analyse zeigt auffällige Messzeitpunkte, beweist jedoch noch keinen technischen Defekt. Für eine genaue Ursache wären zusätzliche Informationen zu Abschaltungen, Wartungen und technischen Störungen erforderlich.

In [132]:
# Alle auffälligen Messzeitpunkte anzeigen

auffaellige_werte = hohe_einstrahlung_df.loc[
    hohe_einstrahlung_df["leistungseinbruch"],
    [
        "anlage",
        "datum",
        "sonneneinstrahlung",
        "einstrahlung_relativ",
        "ac_leistung",
        "ac_leistung_relativ"
    ]
].sort_values(["anlage", "datum"])

auffaellige_werte.round(3)

C:\Users\AnaStefaniuk\AppData\Local\Temp\ipykernel_14988\624398958.py:15: UserWarning: obj.round has no effect with datetime, timedelta, or period dtypes. Use obj.dt.round(...) instead.
  auffaellige_werte.round(3)


,anlage,datum,sonneneinstrahlung,einstrahlung_relativ,ac_leistung,ac_leistung_relativ
3581,Anlage 2,2020-05-19 10:15:00,0.776,0.706,3355.366,0.129
3675,Anlage 2,2020-05-20 10:00:00,0.748,0.680,7785.320,0.300
3676,Anlage 2,2020-05-20 10:15:00,0.864,0.786,5346.840,0.206
3677,Anlage 2,2020-05-20 10:30:00,0.849,0.773,3496.647,0.135
3678,Anlage 2,2020-05-20 10:45:00,0.912,0.830,3669.427,0.141
3679,Anlage 2,2020-05-20 11:00:00,0.878,0.799,3557.437,0.137
3680,Anlage 2,2020-05-20 11:15:00,0.900,0.819,3680.076,0.142
3681,Anlage 2,2020-05-20 11:30:00,0.904,0.823,3701.473,0.142
3682,Anlage 2,2020-05-20 11:45:00,0.893,0.812,3843.013,0.148
3683,Anlage 2,2020-05-20 12:00:00,0.889,0.809,5127.687,0.197


## Auffälligkeiten pro Tag zählen

In [133]:
# Anzahl der Leistungseinbrüche pro Tag berechnen

einbrueche_pro_tag = (
    auffaellige_werte
    .groupby(auffaellige_werte["datum"].dt.date)
    .size()
    .reset_index(name="anzahl_leistungseinbrueche")
)

einbrueche_pro_tag

,datum,anzahl_leistungseinbrueche
0,2020-05-19,1
1,2020-05-20,17
2,2020-05-21,1
3,2020-05-28,2
4,2020-06-02,2
5,2020-06-03,1
6,2020-06-07,1
7,2020-06-09,1
8,2020-06-10,2
9,2020-06-13,2


### Untersuchung der auffälligen Messzeitpunkte

Die Detailprüfung zeigt, dass alle 30 erkannten Leistungseinbrüche bei Anlage 2 auftraten.

Besonders auffällig ist der 20.05.2020. Zwischen 10:00 und 14:00 Uhr blieb die relative AC-Leistung trotz hoher Sonneneinstrahlung durchgehend unter 30 %. Es handelt sich daher nicht nur um einen einzelnen Messwert, sondern um einen länger anhaltenden Leistungseinbruch.

Am 02.06.2020 um 12:30 Uhr sowie am 10.06.2020 um 11:15 und 11:30 Uhr wurde trotz hoher Sonneneinstrahlung überhaupt keine AC-Leistung gemessen. Daneben sind mehrere einzelne Leistungseinbrüche an weiteren Tagen zu erkennen.

Die zeitliche Häufung deutet darauf hin, dass es sich nicht ausschließlich um zufällige Schwankungen handelt. Mögliche Ursachen könnten technische Störungen, Abschaltungen oder Fehler bei der Datenerfassung sein. Eine eindeutige Ursache lässt sich mit den vorhandenen Daten jedoch nicht feststellen.

## Analyse 5: Stromproduktion im Tagesverlauf

### Hypothese

Die höchste Stromproduktion wird bei beiden PV-Anlagen ungefähr zur Mittagszeit erreicht.

## Durchschnittliche Leistung je Stunde berechnen

In [134]:
# Durchschnittliche relative AC-Leistung je Anlage und Stunde berechnen

stundenleistung_df = zeitpunkt_df.groupby(["anlage", "stunde"],as_index=False)["ac_leistung_relativ"].mean()

In [135]:
# Ergebnisspalte umbenennen

stundenleistung_df = stundenleistung_df.rename(columns={"ac_leistung_relativ": "durchschnittliche_ac_leistung"})

In [136]:
# Durchschnittliche Leistung auf drei Nachkommastellen runden

stundenleistung_df["durchschnittliche_ac_leistung"] = (stundenleistung_df["durchschnittliche_ac_leistung"].round(3))

stundenleistung_df.head()

,anlage,stunde,durchschnittliche_ac_leistung
0,Anlage 1,0,0.0
1,Anlage 1,1,0.0
2,Anlage 1,2,0.0
3,Anlage 1,3,0.0
4,Anlage 1,4,0.0


## Stunde mit der höchsten Leistung bestimmen

In [137]:
# Stunden nach Anlage und durchschnittlicher Leistung sortieren

stundenleistung_sortiert = stundenleistung_df.sort_values(["anlage", "durchschnittliche_ac_leistung"],ascending=[True, False])

In [138]:
# Stunde mit der höchsten durchschnittlichen Leistung je Anlage anzeigen

hoechste_stundenleistung = stundenleistung_sortiert.groupby("anlage").head(1)

hoechste_stundenleistung

,anlage,stunde,durchschnittliche_ac_leistung
11,Anlage 1,11,0.723
36,Anlage 2,12,0.524


## Tagesverlauf als Liniendiagramm darstellen

In [139]:
# Durchschnittlichen Tagesverlauf beider Anlagen darstellen

fig = px.line(
    stundenleistung_df,
    x="stunde",
    y="durchschnittliche_ac_leistung",
    color="anlage",
    color_discrete_map=FARBEN,
    markers=True,
    title="Durchschnittliche Stromproduktion im Tagesverlauf",
    labels={
        "stunde": "Stunde",
        "durchschnittliche_ac_leistung": "Durchschnittliche relative AC-Leistung",
        "anlage": "PV-Anlage"
    }
)

# Alle Stunden von 0 bis 23 auf der X-Achse anzeigen
fig.update_xaxes(
    tickmode="linear",
    dtick=1
)

fig.show()

### Ergebnis

Die Hypothese wird bestätigt. Beide PV-Anlagen erreichen ihre höchste durchschnittliche AC-Leistung ungefähr zur Mittagszeit.

Bei Anlage 1 wird der höchste Durchschnittswert um 11 Uhr erreicht. Die relative AC-Leistung beträgt zu dieser Stunde durchschnittlich 72,3 % des Anlagenhöchstwertes.

Bei Anlage 2 liegt der höchste Durchschnittswert um 12 Uhr bei 52,4 % des Anlagenhöchstwertes.

Der niedrigere Wert von Anlage 2 passt zu den vorherigen Ergebnissen. Bei dieser Anlage wurden größere Leistungsunterschiede zwischen den Wechselrichtern und mehrere Leistungseinbrüche festgestellt.

Nach dem mittäglichen Höchstwert nimmt die durchschnittliche Leistung beider Anlagen im weiteren Tagesverlauf wieder ab.

# Gesamtergebnisse

Die Analyse untersuchte den Einfluss von Sonneneinstrahlung und Temperatur auf die Stromproduktion von zwei Photovoltaikanlagen. Zusätzlich wurden die Leistung einzelner Wechselrichter sowie auffällige Leistungseinbrüche analysiert.

## Wichtigste Ergebnisse

1. **Sonneneinstrahlung und AC-Leistung**

   Bei beiden Anlagen besteht ein sehr starker positiver Zusammenhang zwischen Sonneneinstrahlung und AC-Leistung. Die Korrelation beträgt bei Anlage 1 rund 1,00 und bei Anlage 2 rund 0,91.

2. **Temperatur und AC-Leistung**

   Die Modultemperatur hängt bei beiden Anlagen stärker mit der AC-Leistung zusammen als die Außentemperatur. Bei Anlage 1 beträgt die Korrelation 0,96 und bei Anlage 2 0,87. Dieser Zusammenhang entsteht jedoch auch dadurch, dass starke Sonneneinstrahlung gleichzeitig die Module erwärmt und die Stromproduktion erhöht.

3. **Unterschiede zwischen den Wechselrichtern**

   Bei Anlage 1 liegen die auffälligsten Wechselrichter 9,15 % und 7,70 % unter dem Anlagendurchschnitt.

   Bei Anlage 2 sind die Unterschiede deutlich größer. Die drei schwächsten Wechselrichter liegen 23,52 %, 21,76 % und 19,01 % unter dem Anlagendurchschnitt.

4. **Auffällige Leistungseinbrüche**

   Bei Anlage 1 wurde bei hoher Sonneneinstrahlung kein auffälliger Leistungseinbruch festgestellt.

   Bei Anlage 2 wurden 30 auffällige Messzeitpunkte erkannt. Das entspricht 5,78 % der untersuchten Messungen mit hoher Sonneneinstrahlung.

   Besonders auffällig war der 20.05.2020. Zwischen 10:00 und 14:00 Uhr blieb die AC-Leistung trotz hoher Sonneneinstrahlung deutlich reduziert.

5. **Tagesverlauf**

   Beide Anlagen erreichen ihre höchste durchschnittliche AC-Leistung ungefähr zur Mittagszeit. Anlage 1 erreicht ihren höchsten Durchschnittswert um 11 Uhr und Anlage 2 um 12 Uhr.

# Fazit

Die Sonneneinstrahlung ist der wichtigste untersuchte Einflussfaktor auf die Stromproduktion beider PV-Anlagen. Mit zunehmender Sonneneinstrahlung steigt normalerweise auch die erzeugte AC-Leistung.

Anlage 1 zeigt insgesamt einen gleichmäßigeren Zusammenhang zwischen Sonneneinstrahlung und Stromproduktion. Auch die Erträge ihrer Wechselrichter liegen überwiegend nah beieinander.

Anlage 2 weist dagegen deutlich größere Schwankungen auf. Mehrere Wechselrichter erzeugen wesentlich weniger Energie als der Anlagendurchschnitt. Zusätzlich wurden mehrere Zeiträume erkannt, in denen die AC-Leistung trotz hoher Sonneneinstrahlung ungewöhnlich niedrig war.

Die Hauptfrage kann daher so beantwortet werden: Sonneneinstrahlung und Modultemperatur hängen stark mit der Stromproduktion zusammen. Gleichzeitig lassen sich bei Anlage 2 auffällige Leistungseinbußen und Unterschiede zwischen einzelnen Wechselrichtern erkennen.

# Handlungsempfehlungen

- Die drei auffälligsten Wechselrichter von Anlage 2 sollten technisch überprüft werden.
- Die längeren Leistungseinbrüche, besonders am 20.05.2020, sollten mit Wartungs- und Störungsprotokollen verglichen werden.
- Zeitpunkte mit hoher Sonneneinstrahlung und ungewöhnlich niedriger Leistung sollten automatisch überwacht werden.
- Für zukünftige Analysen sollten zusätzlich Informationen zu Wartungen, Abschaltungen und technischen Fehlern erfasst werden.
- Die Anlagen sollten weiterhin getrennt bewertet werden, weil ihre Leistungswerte und technischen Eigenschaften unterschiedlich sind.

# Einschränkungen der Analyse

- Der Datensatz umfasst nur den Zeitraum vom 15.05.2020 bis zum 17.06.2020.
- Angaben zu Standort, installierter Leistung und technischen Eigenschaften der Anlagen fehlen.
- Die Ursachen der Leistungseinbrüche sind nicht im Datensatz enthalten.
- Die unterschiedlichen Wertebereiche der beiden Anlagen erlauben keinen direkten Vergleich der absoluten Leistung.
- Die verwendeten Grenzen für einen Leistungseinbruch wurden für diese Analyse selbst festgelegt.
- Korrelationen zeigen Zusammenhänge, beweisen aber keine Ursache.

## Leistungseinbrüche speichern

In [140]:
# Alle auffälligen Messzeitpunkte speichern

auffaellige_werte.to_csv("auffaellige_leistungseinbrueche.csv",index=False)

## Prüfen, ob die Dateien erstellt wurden

In [141]:
# Inhalt des aktuellen Notebook-Ordners anzeigen

import os

os.listdir()

['.git',
 '.gitignore.txt',
 'auffaellige_leistungseinbrueche.csv',
 'main.ipynb',
 'Plant_1_Generation_Data.csv',
 'Plant_1_Weather_Sensor_Data.csv',
 'Plant_2_Generation_Data.csv',
 'Plant_2_Weather_Sensor_Data.csv',
 'pv_analyse_bereinigt.csv',
 'pv_zeitpunkte.csv',
 'README.md',
 'requirements.txt',
 'wechselrichter_ergebnisse.csv']

In [142]:
# Alle wichtigen Ergebnistabellen speichern

analyse_df.to_csv(
    "pv_analyse_bereinigt.csv",
    index=False
)

zeitpunkt_df.to_csv(
    "pv_zeitpunkte.csv",
    index=False
)

wechselrichter_ertrag_df.to_csv(
    "wechselrichter_ergebnisse.csv",
    index=False
)

auffaellige_werte.to_csv(
    "auffaellige_leistungseinbrueche.csv",
    index=False
)

print("Alle Dateien wurden gespeichert.")

Alle Dateien wurden gespeichert.


In [143]:
# Nur die erzeugten Ergebnisdateien anzeigen

[datei for datei in os.listdir()
    if datei.startswith(("pv_", "wechselrichter_", "auffaellige_"))]

['auffaellige_leistungseinbrueche.csv',
 'pv_analyse_bereinigt.csv',
 'pv_zeitpunkte.csv',
 'wechselrichter_ergebnisse.csv']

In [144]:
# Alle Ergebnisdateien speichern beziehungsweise aktualisieren

# 1. Vollständige bereinigte Analysetabelle
analyse_df.to_csv(
    "pv_analyse_bereinigt.csv",
    index=False
)

# 2. Zusammengefasste Werte je Anlage und Zeitpunkt
zeitpunkt_df.to_csv(
    "pv_zeitpunkte.csv",
    index=False
)

# 3. Wechselrichter-Ergebnisse mit Datenabdeckung
wechselrichter_ertrag_df.to_csv(
    "wechselrichter_ergebnisse.csv",
    index=False
)

# 4. Ursprünglich erkannte 30 Leistungseinbrüche
auffaellige_werte.to_csv(
    "auffaellige_leistungseinbrueche.csv",
    index=False
)

# 5. Korrigierte Leistungseinbrüche auswählen
auffaellige_werte_korrigiert = (
    hohe_einstrahlung_korrigiert_df.loc[
        hohe_einstrahlung_korrigiert_df[
            "leistungseinbruch_korrigiert"
        ]
    ]
    .sort_values(["anlage", "datum"])
)

# 6. Korrigierte 19 Leistungseinbrüche speichern
auffaellige_werte_korrigiert.to_csv(
    "auffaellige_leistungseinbrueche_korrigiert.csv",
    index=False
)

print("Alle Ergebnisdateien wurden erfolgreich gespeichert.")

Alle Ergebnisdateien wurden erfolgreich gespeichert.
